

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>haploid_highd Overview &mdash; FFPopSim 1.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="FFPopSim 1.1 documentation" href="index.html" />
    <link rel="up" title="Contents" href="Contents.html" />
    <link rel="next" title="hivpopulation Overview" href="hivpopulation_categories.html" />
    <link rel="prev" title="haploid_lowd Overview" href="haploid_lowd_categories.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="hivpopulation_categories.html" title="hivpopulation Overview"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="haploid_lowd_categories.html" title="haploid_lowd Overview"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">FFPopSim 1.1 documentation</a> &raquo;</li>
          <li><a href="Contents.html" accesskey="U">Contents</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="haploid-highd-overview">
<h1>haploid_highd Overview<a class="headerlink" href="#haploid-highd-overview" title="Permalink to this headline">¶</a></h1>
<p>The following lists all methods and attributes of <tt class="docutils literal"><span class="pre">haploid_highd</span></tt>, grouped according to purpose and function.
Member functions are shown with the prefix <tt class="docutils literal"><span class="pre">haploid_highd.*</span></tt>.
If you have initialized for example as <tt class="docutils literal"><span class="pre">pop=haploid_highd(L)</span></tt>, the prefix in your program needs to be <tt class="docutils literal"><span class="pre">pop.*</span></tt>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Clicking on the name of the function will take you to a more detailed explanation listing all arguments.</p>
</div>
<div class="section" id="initialization">
<h2>Initialization<a class="headerlink" href="#initialization" title="Permalink to this headline">¶</a></h2>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.__init__" title="FFPopSim.haploid_highd.__init__"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.__init__</span></tt></a></td>
<td>Construct a high-dimensional population with certain parameters.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="attributes">
<h2>Attributes<a class="headerlink" href="#attributes" title="Permalink to this headline">¶</a></h2>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.L" title="FFPopSim.haploid_highd.L"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.L</span></tt></a></td>
<td>Number of loci (read-only)</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.number_of_loci" title="FFPopSim.haploid_highd.number_of_loci"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.number_of_loci</span></tt></a></td>
<td>Number of loci (read-only)</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.N" title="FFPopSim.haploid_highd.N"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.N</span></tt></a></td>
<td>Population size (read-only)</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.population_size" title="FFPopSim.haploid_highd.population_size"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.population_size</span></tt></a></td>
<td>Population size (read-only)</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.number_of_traits" title="FFPopSim.haploid_highd.number_of_traits"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.number_of_traits</span></tt></a></td>
<td>Number of traits (read-only)</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.generation" title="FFPopSim.haploid_highd.generation"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.generation</span></tt></a></td>
<td>Current generation (read-only)</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.participation_ratio" title="FFPopSim.haploid_highd.participation_ratio"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.participation_ratio</span></tt></a></td>
<td>Participation ratio (read-only)</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.number_of_clones" title="FFPopSim.haploid_highd.number_of_clones"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.number_of_clones</span></tt></a></td>
<td>Number of clones (read-only)</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.carrying_capacity" title="FFPopSim.haploid_highd.carrying_capacity"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.carrying_capacity</span></tt></a></td>
<td>current carrying capacity of the environment</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.circular" title="FFPopSim.haploid_highd.circular"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.circular</span></tt></a></td>
<td>is the genome circular?</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.recombination_model" title="FFPopSim.haploid_highd.recombination_model"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.recombination_model</span></tt></a></td>
<td>model of recombination to use</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.outcrossing_rate" title="FFPopSim.haploid_highd.outcrossing_rate"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.outcrossing_rate</span></tt></a></td>
<td>outcrossing rate (probability of sexual reproduction per generation)</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.crossover_rate" title="FFPopSim.haploid_highd.crossover_rate"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.crossover_rate</span></tt></a></td>
<td>crossover rate (probability of crossover per site per generation)</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.mutation_rate" title="FFPopSim.haploid_highd.mutation_rate"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.mutation_rate</span></tt></a></td>
<td>mutation rate (per site per generation)</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="initialize-the-population">
<h2>Initialize the population<a class="headerlink" href="#initialize-the-population" title="Permalink to this headline">¶</a></h2>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.set_wildtype" title="FFPopSim.haploid_highd.set_wildtype"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.set_wildtype</span></tt></a></td>
<td>Set up a population of wildtype individuals</td>
</tr>
<tr class="row-even"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.set_allele_frequencies</span></tt></td>
<td>Initialize the population according to the given allele frequencies in linkage equilibrium.</td>
</tr>
<tr class="row-odd"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.set_genotypes</span></tt></td>
<td>Initialize population with fixed counts for specific genotypes.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.add_genotypes" title="FFPopSim.haploid_highd.add_genotypes"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.add_genotypes</span></tt></a></td>
<td>Add new individuals to the population with certain genotypes</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="set-phenotypes-and-fitness-function">
<h2>Set phenotypes and fitness function<a class="headerlink" href="#set-phenotypes-and-fitness-function" title="Permalink to this headline">¶</a></h2>
<p>The following methods are shortcuts if you have only one trait
(true if you do not subclass <tt class="docutils literal"><span class="pre">haploid_highd</span></tt>) and will not work
for populations with more than one phenotypic trait:</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.set_fitness_additive" title="FFPopSim.haploid_highd.set_fitness_additive"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.set_fitness_additive</span></tt></a></td>
<td>Shortcut for set_trait_additive when there is only one trait</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.add_fitness_coefficient" title="FFPopSim.haploid_highd.add_fitness_coefficient"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.add_fitness_coefficient</span></tt></a></td>
<td>Shortcut for add_trait_coefficient when there is only one trait</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.set_random_epistasis" title="FFPopSim.haploid_highd.set_random_epistasis"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.set_random_epistasis</span></tt></a></td>
<td>Shortcut for set_random_trait_epistasis when there is only one trait</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.clear_fitness" title="FFPopSim.haploid_highd.clear_fitness"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.clear_fitness</span></tt></a></td>
<td>Shortcut for clear_trait when there is only one trait</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="evolution">
<h2>Evolution<a class="headerlink" href="#evolution" title="Permalink to this headline">¶</a></h2>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.evolve</span></tt></td>
<td>Evolve for some generations.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.bottleneck" title="FFPopSim.haploid_highd.bottleneck"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.bottleneck</span></tt></a></td>
<td>Make the population undergo a bottleneck</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.unique_clones" title="FFPopSim.haploid_highd.unique_clones"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.unique_clones</span></tt></a></td>
<td>Recompress the clone structure</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.calc_stat" title="FFPopSim.haploid_highd.calc_stat"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.calc_stat</span></tt></a></td>
<td>Calculate trait and fitness statistics for the population</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="random-sampling">
<h2>Random Sampling<a class="headerlink" href="#random-sampling" title="Permalink to this headline">¶</a></h2>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.random_clone" title="FFPopSim.haploid_highd.random_clone"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.random_clone</span></tt></a></td>
<td>Get a random clone</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.random_clones" title="FFPopSim.haploid_highd.random_clones"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.random_clones</span></tt></a></td>
<td>Get random clones</td>
</tr>
<tr class="row-odd"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.random_genomes</span></tt></td>
<td>Get random genomes from the population</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="clone-structure-and-genotypes">
<h2>Clone structure and genotypes<a class="headerlink" href="#clone-structure-and-genotypes" title="Permalink to this headline">¶</a></h2>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_clone_size" title="FFPopSim.haploid_highd.get_clone_size"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_clone_size</span></tt></a></td>
<td>Get the size of a clone</td>
</tr>
<tr class="row-even"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_clone_sizes</span></tt></td>
<td>Get the fitness of all clones.</td>
</tr>
<tr class="row-odd"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_genotype</span></tt></td>
<td>Get a genotype from the population</td>
</tr>
<tr class="row-even"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_genotypes</span></tt></td>
<td>Get genotypes of the population.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="get-allele-frequencies-and-linkage-disequilibria">
<h2>Get allele frequencies and linkage disequilibria<a class="headerlink" href="#get-allele-frequencies-and-linkage-disequilibria" title="Permalink to this headline">¶</a></h2>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_allele_frequency" title="FFPopSim.haploid_highd.get_allele_frequency"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_allele_frequency</span></tt></a></td>
<td>Get the frequency of the + allele at the selected locus</td>
</tr>
<tr class="row-even"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_allele_frequencies</span></tt></td>
<td>Get all allele frequencies</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_pair_frequency" title="FFPopSim.haploid_highd.get_pair_frequency"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_pair_frequency</span></tt></a></td>
<td>Get the joint frequency of two + alleles</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_chi" title="FFPopSim.haploid_highd.get_chi"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_chi</span></tt></a></td>
<td>Get chi of an allele in the -/+ basis</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_chi2" title="FFPopSim.haploid_highd.get_chi2"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_chi2</span></tt></a></td>
<td>Get <span class="math">\(\chi_{ij}\)</span></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_LD" title="FFPopSim.haploid_highd.get_LD"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_LD</span></tt></a></td>
<td>Get linkage disequilibrium</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_moment" title="FFPopSim.haploid_highd.get_moment"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_moment</span></tt></a></td>
<td>Get moment of two alleles in the -/+ basis</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="analyze-fitness-and-trait-distributions">
<h2>Analyze fitness and trait distributions<a class="headerlink" href="#analyze-fitness-and-trait-distributions" title="Permalink to this headline">¶</a></h2>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.calc_stat" title="FFPopSim.haploid_highd.calc_stat"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.calc_stat</span></tt></a></td>
<td>Calculate trait and fitness statistics for the population</td>
</tr>
<tr class="row-even"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_trait_additive</span></tt></td>
<td>Get an array with the additive coefficients of all loci of a trait.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_trait_statistics" title="FFPopSim.haploid_highd.get_trait_statistics"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_trait_statistics</span></tt></a></td>
<td>Get the mean and variance of a trait in the population.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_trait_covariance" title="FFPopSim.haploid_highd.get_trait_covariance"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_trait_covariance</span></tt></a></td>
<td>Get the covariance of two traits in the population.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_fitness" title="FFPopSim.haploid_highd.get_fitness"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_fitness</span></tt></a></td>
<td>Get the fitness of an individual</td>
</tr>
<tr class="row-even"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_fitnesses</span></tt></td>
<td>Get the fitness of all clones.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_fitness_statistics" title="FFPopSim.haploid_highd.get_fitness_statistics"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_fitness_statistics</span></tt></a></td>
<td>Get the mean and variance of the fitness in the population.</td>
</tr>
<tr class="row-even"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_fitness_histogram</span></tt></td>
<td>Calculate the fitness histogram.</td>
</tr>
<tr class="row-odd"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.plot_fitness_histogram</span></tt></td>
<td>Plot a distribution of fitness in the population.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="divergence-and-diversity">
<h2>Divergence and Diversity<a class="headerlink" href="#divergence-and-diversity" title="Permalink to this headline">¶</a></h2>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_divergence_statistics" title="FFPopSim.haploid_highd.get_divergence_statistics"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_divergence_statistics</span></tt></a></td>
<td>Get the mean and variance of the divergence in the population.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_diversity_statistics" title="FFPopSim.haploid_highd.get_diversity_statistics"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_diversity_statistics</span></tt></a></td>
<td>Get the mean and variance of the diversity in the population.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="haploid_highd.html#FFPopSim.haploid_highd.get_trait_statistics" title="FFPopSim.haploid_highd.get_trait_statistics"><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_trait_statistics</span></tt></a></td>
<td>Get the mean and variance of a trait in the population.</td>
</tr>
<tr class="row-even"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_divergence_histogram</span></tt></td>
<td>Get the divergence histogram restricted to those chunks of the genome.</td>
</tr>
<tr class="row-odd"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.get_diversity_histogram</span></tt></td>
<td>Get the diversity histogram restricted to those chunks of the genome.</td>
</tr>
<tr class="row-even"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.plot_divergence_histogram</span></tt></td>
<td>Plot the divergence histogram.</td>
</tr>
<tr class="row-odd"><td><tt class="xref py py-obj docutils literal"><span class="pre">haploid_highd.plot_diversity_histogram</span></tt></td>
<td>Plot the diversity histogram.</td>
</tr>
</tbody>
</table>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">haploid_highd Overview</a><ul>
<li><a class="reference internal" href="#initialization">Initialization</a></li>
<li><a class="reference internal" href="#attributes">Attributes</a></li>
<li><a class="reference internal" href="#initialize-the-population">Initialize the population</a></li>
<li><a class="reference internal" href="#set-phenotypes-and-fitness-function">Set phenotypes and fitness function</a></li>
<li><a class="reference internal" href="#evolution">Evolution</a></li>
<li><a class="reference internal" href="#random-sampling">Random Sampling</a></li>
<li><a class="reference internal" href="#clone-structure-and-genotypes">Clone structure and genotypes</a></li>
<li><a class="reference internal" href="#get-allele-frequencies-and-linkage-disequilibria">Get allele frequencies and linkage disequilibria</a></li>
<li><a class="reference internal" href="#analyze-fitness-and-trait-distributions">Analyze fitness and trait distributions</a></li>
<li><a class="reference internal" href="#divergence-and-diversity">Divergence and Diversity</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="haploid_lowd_categories.html"
                        title="previous chapter">haploid_lowd Overview</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="hivpopulation_categories.html"
                        title="next chapter">hivpopulation Overview</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/haploid_highd_categories.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="hivpopulation_categories.html" title="hivpopulation Overview"
             >next</a> |</li>
        <li class="right" >
          <a href="haploid_lowd_categories.html" title="haploid_lowd Overview"
             >previous</a> |</li>
        <li><a href="index.html">FFPopSim 1.1 documentation</a> &raquo;</li>
          <li><a href="Contents.html" >Contents</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012, Fabio Zanini, Richard Neher.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>