

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>haploid_lowd Class Reference &mdash; FFPopSim 1.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="FFPopSim 1.1 documentation" href="index.html" />
    <link rel="up" title="Contents" href="Contents.html" />
    <link rel="next" title="haploid_highd Class Reference" href="haploid_highd.html" />
    <link rel="prev" title="hivpopulation Overview" href="hivpopulation_categories.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="haploid_highd.html" title="haploid_highd Class Reference"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="hivpopulation_categories.html" title="hivpopulation Overview"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">FFPopSim 1.1 documentation</a> &raquo;</li>
          <li><a href="Contents.html" accesskey="U">Contents</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="haploid-lowd-class-reference">
<span id="haploid-lowd"></span><h1>haploid_lowd Class Reference<a class="headerlink" href="#haploid-lowd-class-reference" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="FFPopSim.haploid_lowd">
<em class="property">class </em><tt class="descname">haploid_lowd</tt><big>(</big><em>L=1</em>, <em>rng_seed=0</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd" title="Permalink to this definition">¶</a></dt>
<dd><p>Class for low-dimensional population genetics (short genomes ~20 loci).</p>
<p>The class offers a number of functions, but an example will explain the basic idea:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c">#####################################</span>
<span class="c">#   EXAMPLE SCRIPT                  #</span>
<span class="c">#####################################</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">FFPopSim</span> <span class="kn">as</span> <span class="nn">h</span>

<span class="n">c</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">haploid_lowd</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>               <span class="c"># 5 loci</span>

<span class="c"># initialize with 300 individuals with genotype 00000,</span>
<span class="c"># and 700 with genotype 00010</span>
<span class="n">c</span><span class="o">.</span><span class="n">set_genotypes</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">300</span><span class="p">,</span> <span class="mi">700</span><span class="p">])</span>

<span class="c"># set an additive fitness landscape with these coefficients</span>
<span class="n">c</span><span class="o">.</span><span class="n">set_fitness_additive</span><span class="p">([</span><span class="mf">0.02</span><span class="p">,</span><span class="mf">0.03</span><span class="p">,</span><span class="mf">0.04</span><span class="p">,</span><span class="mf">0.02</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.03</span><span class="p">])</span>
<span class="c"># Note: we are in the -/+ basis, so</span>
<span class="c">#        F[10000] - F[00000] = 2 * 0.02 </span>
<span class="c"># Hence the coefficients are half of the effect of mutation on fitness </span>

<span class="n">c</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>                       <span class="c"># evolve for 100 generations</span>
<span class="n">c</span><span class="o">.</span><span class="n">plot_diversity_histogram</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="c">#####################################</span>
</pre></div>
</div>
<dl class="method">
<dt id="FFPopSim.haploid_lowd.__init__">
<tt class="descname">__init__</tt><big>(</big><em>L=1</em>, <em>rng_seed=0</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct a low-dimensional population with certain parameters.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>L : number of loci (at least 1)</li>
<li>rng_seed : seed for the random number generator</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.allele_entropy">
<tt class="descname">allele_entropy</tt><big>(</big><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.allele_entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>get the allele entropy of the population</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the allele entropy is defined as <span class="math">\(-\sum_{i=0}^{L} \nu_i\log \nu_i + (1-\nu_i)\log(1-\nu_i)\)</span>.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.evolve">
<tt class="descname">evolve</tt><big>(</big><em>gen=1</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.evolve" title="Permalink to this definition">¶</a></dt>
<dd><p>Evolve for some generations</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>gen: number of generations to evolve the population, defaults to one</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.evolve_deterministic">
<tt class="descname">evolve_deterministic</tt><big>(</big><em>gen=1</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.evolve_deterministic" title="Permalink to this definition">¶</a></dt>
<dd><p>Evolve for some generations deterministically (skips the resampling)</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>gen: number of generations to evolve the population</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.evolve_norec">
<tt class="descname">evolve_norec</tt><big>(</big><em>gen=1</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.evolve_norec" title="Permalink to this definition">¶</a></dt>
<dd><p>Evolve for some generations without recombination</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>gen: number of generations to evolve the population</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.genotype_entropy">
<tt class="descname">genotype_entropy</tt><big>(</big><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.genotype_entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the genotype entropy of the population</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the genotype entropy is defined as <span class="math">\(-\sum_{i=0}^{2^L} p_i \log p_i\)</span>.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_allele_frequency">
<tt class="descname">get_allele_frequency</tt><big>(</big><em>locus</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_allele_frequency" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the frequency of the + allele</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>locus: locus, at which the frequency of the + allele is to be computed</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>the frequency of the + allele, <span class="math">\(\nu_i := \frac{1 + \left&lt;s_i\right&gt;}{2}\)</span>, where <span class="math">\(s_i \in \{-1, 1\}\)</span>.</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_pair_frequency">
<tt class="descname">get_pair_frequency</tt><big>(</big><em>locus1</em>, <em>locus2</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_pair_frequency" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the joint frequency of two + alleles</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>locus1: first locus</li>
<li>locus2: second locus</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>the joint frequency of the + alleles</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_chi">
<tt class="descname">get_chi</tt><big>(</big><em>locus</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_chi" title="Permalink to this definition">¶</a></dt>
<dd><p>Get chi of an allele in the -/+ basis</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>locus: locus whose chi is to be computed</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>the chi of that allele, <span class="math">\(\chi_i := \left&lt;s_i\right&gt;\)</span>, where <span class="math">\(s_i \in \{-1, 1\}\)</span>.</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_chi2">
<tt class="descname">get_chi2</tt><big>(</big><em>locus1</em>, <em>locus2</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_chi2" title="Permalink to this definition">¶</a></dt>
<dd><p>Get <span class="math">\(\chi_{ij}\)</span></p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>locus1: first locus</li>
<li>locus2: second locus</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>the linkage disequilibiurm between them, i.e. <span class="math">\(\chi_{ij} := \left&lt;s_i s_j\right&gt; - \chi_i \cdot \chi_j\)</span>.</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_fitness">
<tt class="descname">get_fitness</tt><big>(</big><em>genotype</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_fitness" title="Permalink to this definition">¶</a></dt>
<dd><p>Get fitness values of a genotype</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>gt: genotype whose fitness is to be calculated. This can either be an integer or in binary format, e.g. 5 = 0b101</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>the fitness of that genotype.</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_genotype_frequency">
<tt class="descname">get_genotype_frequency</tt><big>(</big><em>locus</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_genotype_frequency" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the frequency of a genotype</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>gt: genotype, whose the frequency is to be returned</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>the frequency of the genotype</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_LD">
<tt class="descname">get_LD</tt><big>(</big><em>locus1</em>, <em>locus2</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_LD" title="Permalink to this definition">¶</a></dt>
<dd><p>Get linkage disequilibrium</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>locus1: first locus</li>
<li>locus2: second locus</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>the linkage disequilibiurm between them, i.e. <span class="math">\(LD := 1 / 4 \left&lt;s_i s_j\right&gt; - \chi_i \cdot \chi_j\)</span>.</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_moment">
<tt class="descname">get_moment</tt><big>(</big><em>locus1</em>, <em>locus2</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_moment" title="Permalink to this definition">¶</a></dt>
<dd><p>Get moment of two alleles in the -/+ basis</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>locus1: first locus</li>
<li>locus2: second locus</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>the second moment, i.e. <span class="math">\(\left&lt;s_i s_j\right&gt;\)</span>, where <span class="math">\(s_i, s_j \in \{-1, 1\}\)</span>.</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.set_fitness_additive">
<tt class="descname">set_fitness_additive</tt><big>(</big><em>coefficient</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.set_fitness_additive" title="Permalink to this definition">¶</a></dt>
<dd><p>Set an additive fitness landscape</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>coefficients: array/list of additive fitness coefficients. It must have length L.</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.set_recombination_rates">
<tt class="descname">set_recombination_rates</tt><big>(</big><em>rec_rates</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.set_recombination_rates" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the recombination rate(s).</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>rates: if a double, the recombination rate at between any two loci; if an array,
the locus-specific recombination rates</li>
<li>model: the recombination model to use (CROSSOVERS or, for linear
genomes, SINGLE_CROSSOVER)</li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">if locus-specific rates are specified, the array must have length
(L-1) for linear chromosomes and length L for circular ones. The
i-th element is the crossover rate between the i-th site and the
(i+1)-th site.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.set_wildtype">
<tt class="descname">set_wildtype</tt><big>(</big><em>N</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.set_wildtype" title="Permalink to this definition">¶</a></dt>
<dd><p>Set up a population of wildtype individuals</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>N: the number of individuals</li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the carrying capacity is set to the same value if still unset.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_allele_frequencies">
<tt class="descname">get_allele_frequencies</tt><big>(</big><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_allele_frequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the frequencies of all + alleles</p>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_fitnesses">
<tt class="descname">get_fitnesses</tt><big>(</big><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_fitnesses" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the fitness of all possible genotypes.</p>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_genotype_frequencies">
<tt class="descname">get_genotype_frequencies</tt><big>(</big><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_genotype_frequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the frequency of each genotype.</p>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_mutation_rates">
<tt class="descname">get_mutation_rates</tt><big>(</big><em>locus=None</em>, <em>direction=None</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_mutation_rates" title="Permalink to this definition">¶</a></dt>
<dd><p>Get one or several mutation rates.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>locus: get only the mutation rate(s) of this locus</li>
<li>direction: get only the forward or backward mutation rate(s)</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>the mutation rate(s) requensted</li>
</ul>
</dd>
</dl>
<p><strong>Note</strong>: if the mutation rates for all loci and/or directions are the same,
this function will try to be smart and give you the answer you are looking for.
In case of doubt, you will get a matrix (L x 2) with the full mutation rate
landscape.</p>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.random_genomes">
<tt class="descname">random_genomes</tt><big>(</big><em>n_sample</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.random_genomes" title="Permalink to this definition">¶</a></dt>
<dd><p>Get random genomes according to their frequencies.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>n_sample: number of random genomes to sample</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>integers corresponding to random genomes in the population.</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.set_allele_frequencies">
<tt class="descname">set_allele_frequencies</tt><big>(</big><em>frequencies</em>, <em>N</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.set_allele_frequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the population in linkage equilibrium with allele frequencies.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>frequencies: an array of length L with all allele frequencies</li>
<li>N: set the population size and, if still unset, the carrying
capacity to this value</li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the population size is only used for resampling and has therefore
no effect on the speed of the simulation.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.set_fitness_function">
<tt class="descname">set_fitness_function</tt><big>(</big><em>indices</em>, <em>vals</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.set_fitness_function" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the fitness landscape at single points.</p>
<p>Parameters:
- indices: genotype to which the fitness values will be assigned
- vals: fitness values to assign</p>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.set_genotypes">
<tt class="descname">set_genotypes</tt><big>(</big><em>indices</em>, <em>counts</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.set_genotypes" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize population with fixed counts for specific genotypes.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>indices: list of genotypes to set (e.g. 0 &#8211;&gt; 00...0, L-1 &#8211;&gt; 11...1)</li>
<li>counts: list of counts for those genotypes</li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the population size and, if unset, the carrying capacity will be set as the sum of the counts.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">you can use Python binary notation for the indices, e.g. 0b0110 is 6.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.set_mutation_rates">
<tt class="descname">set_mutation_rates</tt><big>(</big><em>rates</em>, <em>rates_back=None</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.set_mutation_rates" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the mutation rate(s).</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last">
<li><p class="first">rates:if a double, the mutation rate at any locus in both directions
or, if rates_back is not None, only in the forward direction</p>
<p>if a vector, the mutation rate is specified for each locus, the same
in both directions or, if rates_back is not None, only in the
forward direction</p>
</li>
<li><p class="first">rates_back: mutation rate in the backward direction (global or
locus-specific)</p>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_divergence_statistics">
<tt class="descname">get_divergence_statistics</tt><big>(</big><em>n_sample=1000</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_divergence_statistics" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the mean and variance of the divergence in the population.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>n_sample: number of individuals to sample at random from the population</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>stat: structure with mean and variance of divergence in the population</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_divergence_histogram">
<tt class="descname">get_divergence_histogram</tt><big>(</big><em>bins=10</em>, <em>n_sample=1000</em>, <em>**kwargs</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_divergence_histogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the histogram of the divergence in the population.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>bins: number of bins or list of bin edges (passed verbatim to numpy.histogram)</li>
<li>n_sample: number of individual to sample at random from the population</li>
<li>kwargs: further optional keyword arguments to numpy.histograms</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>h: numpy.histogram of divergence in the population</li>
</ul>
</dd>
</dl>
<p><em>Note</em>: to get a normalized histogram, use the <em>density</em> keyword.</p>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.plot_divergence_histogram">
<tt class="descname">plot_divergence_histogram</tt><big>(</big><em>axis=None</em>, <em>n_sample=1000</em>, <em>**kwargs</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.plot_divergence_histogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the histogram of the divergence in the population.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>axis: use an already existing axis for the plot</li>
<li>n_sample: number of individual to sample at random from the population</li>
<li>kwargs: further optional keyword arguments to numpy.histograms</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_diversity_statistics">
<tt class="descname">get_diversity_statistics</tt><big>(</big><em>n_sample=1000</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_diversity_statistics" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the mean and variance of the diversity in the population.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>n_sample: number of individual to sample at random from the population</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>stat: structure with mean and variance of diversity in the population</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_diversity_histogram">
<tt class="descname">get_diversity_histogram</tt><big>(</big><em>bins=10</em>, <em>n_sample=1000</em>, <em>**kwargs</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_diversity_histogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the histogram of the diversity in the population.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>bins: number of bins or list of bin edges (passed verbatim to numpy.histogram)</li>
<li>n_sample: number of individual to sample at random from the population</li>
<li>kwargs: further optional keyword arguments to numpy.histograms</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>h: numpy.histogram of diversity in the population</li>
</ul>
</dd>
</dl>
<p><em>Note</em>: to get a normalized histogram, use the <em>density</em> keyword.</p>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.plot_diversity_histogram">
<tt class="descname">plot_diversity_histogram</tt><big>(</big><em>axis=None</em>, <em>n_sample=1000</em>, <em>**kwargs</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.plot_diversity_histogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the histogram of the diversity in the population.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>axis: use an already existing axis for the plot</li>
<li>n_sample: number of individual to sample at random from the population</li>
<li>kwargs: further optional keyword arguments to numpy.histograms</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_fitness_statistics">
<tt class="descname">get_fitness_statistics</tt><big>(</big><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_fitness_statistics" title="Permalink to this definition">¶</a></dt>
<dd><p>haploid_lowd_get_fitness_statistics(haploid_lowd self) -&gt; stat</p>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.get_fitness_histogram">
<tt class="descname">get_fitness_histogram</tt><big>(</big><em>n_sample=1000</em>, <em>**kwargs</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.get_fitness_histogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the histogram of the fitness in the population.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>n_sample: number of individual to sample at random from the population</li>
</ul>
</dd>
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>h: numpy.histogram of fitness in the population</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.haploid_lowd.plot_fitness_histogram">
<tt class="descname">plot_fitness_histogram</tt><big>(</big><em>axis=None</em>, <em>n_sample=1000</em>, <em>**kwargs</em><big>)</big><a class="headerlink" href="#FFPopSim.haploid_lowd.plot_fitness_histogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the histogram of the fitness in the population.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>axis: use an already existing axis for the plot</li>
<li>n_sample: number of individual to sample at random from the population</li>
<li>kwargs: further optional keyword arguments to numpy.histograms</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="FFPopSim.haploid_lowd.L">
<tt class="descname">L</tt><a class="headerlink" href="#FFPopSim.haploid_lowd.L" title="Permalink to this definition">¶</a></dt>
<dd><p>number of loci (read-only)</p>
</dd></dl>

<dl class="attribute">
<dt id="FFPopSim.haploid_lowd.number_of_loci">
<tt class="descname">number_of_loci</tt><a class="headerlink" href="#FFPopSim.haploid_lowd.number_of_loci" title="Permalink to this definition">¶</a></dt>
<dd><p>number of loci (read-only)</p>
</dd></dl>

<dl class="attribute">
<dt id="FFPopSim.haploid_lowd.N">
<tt class="descname">N</tt><a class="headerlink" href="#FFPopSim.haploid_lowd.N" title="Permalink to this definition">¶</a></dt>
<dd><p>population size (read-only)</p>
</dd></dl>

<dl class="attribute">
<dt id="FFPopSim.haploid_lowd.population_size">
<tt class="descname">population_size</tt><a class="headerlink" href="#FFPopSim.haploid_lowd.population_size" title="Permalink to this definition">¶</a></dt>
<dd><p>population size (read-only)</p>
</dd></dl>

<dl class="attribute">
<dt id="FFPopSim.haploid_lowd.generation">
<tt class="descname">generation</tt><a class="headerlink" href="#FFPopSim.haploid_lowd.generation" title="Permalink to this definition">¶</a></dt>
<dd><p>current generation (read-only)</p>
</dd></dl>

<dl class="attribute">
<dt id="FFPopSim.haploid_lowd.circular">
<tt class="descname">circular</tt><a class="headerlink" href="#FFPopSim.haploid_lowd.circular" title="Permalink to this definition">¶</a></dt>
<dd><p>is the genome circular?</p>
</dd></dl>

<dl class="attribute">
<dt id="FFPopSim.haploid_lowd.carrying_capacity">
<tt class="descname">carrying_capacity</tt><a class="headerlink" href="#FFPopSim.haploid_lowd.carrying_capacity" title="Permalink to this definition">¶</a></dt>
<dd><p>current carrying capacity of the environment</p>
</dd></dl>

<dl class="attribute">
<dt id="FFPopSim.haploid_lowd.outcrossing_rate">
<tt class="descname">outcrossing_rate</tt><a class="headerlink" href="#FFPopSim.haploid_lowd.outcrossing_rate" title="Permalink to this definition">¶</a></dt>
<dd><p>outcrossing rate (probability of sexual reproduction per generation)</p>
</dd></dl>

</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="hivpopulation_categories.html"
                        title="previous chapter">hivpopulation Overview</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="haploid_highd.html"
                        title="next chapter">haploid_highd Class Reference</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/haploid_lowd.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="haploid_highd.html" title="haploid_highd Class Reference"
             >next</a> |</li>
        <li class="right" >
          <a href="hivpopulation_categories.html" title="hivpopulation Overview"
             >previous</a> |</li>
        <li><a href="index.html">FFPopSim 1.1 documentation</a> &raquo;</li>
          <li><a href="Contents.html" >Contents</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012, Fabio Zanini, Richard Neher.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>