

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>hivpopulation Class Reference &mdash; FFPopSim 1.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="FFPopSim 1.1 documentation" href="index.html" />
    <link rel="up" title="Contents" href="Contents.html" />
    <link rel="next" title="hivgene Class reference" href="hivgene.html" />
    <link rel="prev" title="haploid_highd Class Reference" href="haploid_highd.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="hivgene.html" title="hivgene Class reference"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="haploid_highd.html" title="haploid_highd Class Reference"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">FFPopSim 1.1 documentation</a> &raquo;</li>
          <li><a href="Contents.html" accesskey="U">Contents</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="hivpopulation-class-reference">
<span id="hivpopulation"></span><h1>hivpopulation Class Reference<a class="headerlink" href="#hivpopulation-class-reference" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="FFPopSim.hivpopulation">
<em class="property">class </em><tt class="descname">hivpopulation</tt><big>(</big><em>N=0</em>, <em>rng_seed=0</em>, <em>mutation_rate=3e-05</em>, <em>coinfection_rate=0.01</em>, <em>crossover_rate=0.001</em><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation" title="Permalink to this definition">¶</a></dt>
<dd><p>Class for HIV population genetics (genome size = 10000).</p>
<p>This class is the main object for simulating the evolution of HIV.
The class offers a number of functions, but an example will explain the basic
idea:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c">#####################################</span>
<span class="c">#   EXAMPLE SCRIPT                  #</span>
<span class="c">#####################################</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">FFPopSim</span> <span class="kn">as</span> <span class="nn">h</span>

<span class="n">c</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">hivpopulation</span><span class="p">(</span><span class="mi">2000</span><span class="p">)</span>        <span class="c"># Create a population of 2000 individuals</span>
<span class="n">c</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>                    <span class="c"># Evolve (neutrally) for 100 generations</span>
<span class="n">c</span><span class="o">.</span><span class="n">plot_divergence_histogram</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="c">#####################################</span>
</pre></div>
</div>
<p><strong>This class is a subclass of haploid_high and offers most of its methods.</strong>
In addition to the haploid_highd class, this class offers functions for reading
fitness and drug resistance landscapes from a text file, and to save genomes as
plain text or in compressed NumPy format.</p>
<p>Moreover, there are two phenotypic traits, replication and resistance. Their
relative importance for viral fitness is set by the <tt class="docutils literal"><span class="pre">treatment</span></tt> attribute:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">f</span><span class="p">[</span><span class="n">trait</span><span class="p">]</span> <span class="o">=</span> <span class="n">trait</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">treatment</span> <span class="o">*</span> <span class="n">trait</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<p>By default, <tt class="docutils literal"><span class="pre">treatment</span></tt> is set to zero, to simulate non-treated patients.</p>
<p>The gene structure of HIV is not modelled explicitely, except for a stub of
1000 sites between position 7000 and 8000 to roughly model the _env_ gene.</p>
<dl class="method">
<dt id="FFPopSim.hivpopulation.__init__">
<tt class="descname">__init__</tt><big>(</big><em>N=0</em>, <em>rng_seed=0</em>, <em>mutation_rate=3e-5</em>, <em>coinfection_rate=1e-2</em>, <em>crossover_rate=1e-3</em><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct a HIV population with certain parameters.</p>
<p>Parameters:</p>
<blockquote>
<div><ul class="simple">
<li>N     number of viral particles</li>
<li>rng_seed   seed for the random number generator. If this is 0, time(NULL)+getpid() is used.</li>
<li>mutation_rate      mutation rate in events / generation / site</li>
<li>coinfection_rate   probability of coinfection of the same cell by two viral particles in events / generation</li>
<li>crossover_rate     probability of template switching during coinfection in events / site</li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the genome length is 10000 (see HIVGENOME).</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.hivpopulation.write_genotypes">
<tt class="descname">write_genotypes</tt><big>(</big><em>filename</em>, <em>sample_size</em>, <em>gt_label=''</em>, <em>start=0</em>, <em>length=0</em><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation.write_genotypes" title="Permalink to this definition">¶</a></dt>
<dd><p>Store random genotypes into a plain text file.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>filename: string with the name of the file to store the genotype into</li>
<li>sample_size: how many random genotypes to store</li>
<li>gt_label: common fasta label for the genotypes (e.g. &#8216;HIV-sim&#8217;)</li>
<li>start: if only a portion of the genome is to be stored, start from this position</li>
<li>length: store a chunk from <tt class="docutils literal"><span class="pre">start</span></tt> to this length</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.hivpopulation.read_replication_coefficients">
<tt class="descname">read_replication_coefficients</tt><big>(</big><em>filename</em><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation.read_replication_coefficients" title="Permalink to this definition">¶</a></dt>
<dd><p>Read replication coefficients from a text file</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>filename: string with the name of the file to read the coefficients from</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.hivpopulation.read_resistance_coefficients">
<tt class="descname">read_resistance_coefficients</tt><big>(</big><em>filename</em><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation.read_resistance_coefficients" title="Permalink to this definition">¶</a></dt>
<dd><p>Read resistance coefficients from a text file</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>filename: string with the name of the file to read the coefficients from</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.hivpopulation.write_genotypes_compressed">
<tt class="descname">write_genotypes_compressed</tt><big>(</big><em>filename</em>, <em>sample_size</em>, <em>gt_label=''</em>, <em>start=0</em>, <em>length=0</em><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation.write_genotypes_compressed" title="Permalink to this definition">¶</a></dt>
<dd><p>Store random genotypes into a compressed file.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>filename: string with the name of the file to store the genotype into</li>
<li>sample_size: how many random genotypes to store</li>
<li>gt_label: common fasta label for the genotypes (e.g. &#8220;HIV-sim&#8221;)</li>
<li>start: if only a portion of the genome is to be stored, start from this position</li>
<li>length: store a chunk from <tt class="docutils literal"><span class="pre">start</span></tt> to this length</li>
</ul>
</dd>
</dl>
<p>The genotypes can be read using numpy.load.</p>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.hivpopulation.set_trait_landscape">
<tt class="descname">set_trait_landscape</tt><big>(</big><em>traitnumber=0</em>, <em>lethal_fraction=0.05</em>, <em>deleterious_fraction=0.8</em>, <em>adaptive_fraction=0.01</em>, <em>effect_size_lethal=0.8</em>, <em>effect_size_deleterious=0.1</em>, <em>effect_size_adaptive=0.01</em>, <em>env_fraction=0.1</em>, <em>effect_size_env=0.01</em>, <em>number_epitopes=0</em>, <em>epitope_strength=0.05</em>, <em>number_valleys=0</em>, <em>valley_strength=0.1</em><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation.set_trait_landscape" title="Permalink to this definition">¶</a></dt>
<dd><p>Set HIV trait landscape according to some general parameters.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>lethal_fraction: fraction of lethal sites</li>
<li>deleterious_fraction: fraction of deleterious sites</li>
<li>adaptive_fraction: fraction of beneficial sites</li>
<li>effect_size_lethal: effect of lethal changes</li>
<li>effect_size_deleterious: average effect of deleterious changes</li>
<li>effect_size_adaptive: average effect of beneficial changes</li>
<li>env_fraction: fraction of beneficial sites in env</li>
<li>effect_size_env: average effect of beneficial changes in env</li>
<li>number_epitopes: number of (epistatic) epitopes</li>
<li>epitope_strength: average height of an epitope escape mutation</li>
<li>number_valleys: number of (epistatic) valleys</li>
<li>valley_strength: average depth of a valley</li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the effects of deleterious and beneficial sites are exponentially
distributed, i.e. most of them will still be almost neutral.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">fractions refer to first and second positions only. For instance,
by default, 80% of first and second positions outside env are
deleterious.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the third positions are always neutral (synonymous).</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.hivpopulation.set_replication_landscape">
<tt class="descname">set_replication_landscape</tt><big>(</big><em>lethal_fraction=0.05</em>, <em>deleterious_fraction=0.8</em>, <em>adaptive_fraction=0.01</em>, <em>effect_size_lethal=0.8</em>, <em>effect_size_deleterious=0.1</em>, <em>effect_size_adaptive=0.01</em>, <em>env_fraction=0.1</em>, <em>effect_size_env=0.01</em>, <em>number_epitopes=0</em>, <em>epitope_strength=0.05</em>, <em>number_valleys=0</em>, <em>valley_strength=0.1</em><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation.set_replication_landscape" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the phenotypic landscape for the replication capacity of HIV.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>lethal_fraction: fraction of lethal sites</li>
<li>deleterious_fraction: fraction of deleterious sites</li>
<li>adaptive_fraction: fraction of beneficial sites</li>
<li>effect_size_lethal: effect of lethal changes</li>
<li>effect_size_deleterious: average effect of deleterious changes</li>
<li>effect_size_adaptive: average effect of beneficial changes</li>
<li>env_fraction: fraction of beneficial sites in env</li>
<li>effect_size_env: average effect of beneficial changes in env</li>
<li>number_epitopes: number of (epistatic) epitopes</li>
<li>epitope_strength: average height of an epitope escape mutation</li>
<li>number_valleys: number of (epistatic) valleys</li>
<li>valley_strength: average depth of a valley</li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the effects of deleterious and beneficial sites are exponentially
distributed, i.e. most of them will still be almost neutral.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">fractions refer to first and second positions only. For instance,
by default, 80% of first and second positions outside env are
deleterious.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the third positions are always neutral (synonymous).</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.hivpopulation.set_resistance_landscape">
<tt class="descname">set_resistance_landscape</tt><big>(</big><em>lethal_fraction=0.05</em>, <em>deleterious_fraction=0.8</em>, <em>adaptive_fraction=0.01</em>, <em>effect_size_lethal=0.8</em>, <em>effect_size_deleterious=0.1</em>, <em>effect_size_adaptive=0.01</em>, <em>env_fraction=0.1</em>, <em>effect_size_env=0.01</em>, <em>number_epitopes=0</em>, <em>epitope_strength=0.05</em>, <em>number_valleys=0</em>, <em>valley_strength=0.1</em><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation.set_resistance_landscape" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the phenotypic landscape for the drug resistance of HIV.</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>lethal_fraction: fraction of lethal sites</li>
<li>deleterious_fraction: fraction of deleterious sites</li>
<li>adaptive_fraction: fraction of beneficial sites</li>
<li>effect_size_lethal: effect of lethal changes</li>
<li>effect_size_deleterious: average effect of deleterious changes</li>
<li>effect_size_adaptive: average effect of beneficial changes</li>
<li>env_fraction: fraction of beneficial sites in env</li>
<li>effect_size_env: average effect of beneficial changes in env</li>
<li>number_epitopes: number of (epistatic) epitopes</li>
<li>epitope_strength: average height of an epitope escape mutation</li>
<li>number_valleys: number of (epistatic) valleys</li>
<li>valley_strength: average depth of a valley</li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the effects of deleterious and beneficial sites are exponentially
distributed, i.e. most of them will still be almost neutral.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">fractions refer to first and second positions only. For instance,
by default, 80% of first and second positions outside env are
deleterious.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the third positions are always neutral (synonymous).</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.hivpopulation.get_replication_additive">
<tt class="descname">get_replication_additive</tt><big>(</big><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation.get_replication_additive" title="Permalink to this definition">¶</a></dt>
<dd><p>The additive part of the replication lansdscape.</p>
<dl class="docutils">
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>coefficients: array of additive replication coefficients</li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">the -/+ basis is used throughout the library.
If you are used to the 0/1 basis, keep in mind that
the interaction series-expansion is different.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.hivpopulation.get_resistance_additive">
<tt class="descname">get_resistance_additive</tt><big>(</big><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation.get_resistance_additive" title="Permalink to this definition">¶</a></dt>
<dd><p>The additive part of the resistance lansdscape.</p>
<dl class="docutils">
<dt>Returns:</dt>
<dd><ul class="first last simple">
<li>coefficients: array of additive drug resistance coefficients</li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">the -/+ basis is used throughout the library.
If you are used to the 0/1 basis, keep in mind that
the interaction series-expansion is different.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.hivpopulation.set_replication_additive">
<tt class="descname">set_replication_additive</tt><big>(</big><em>coefficients</em><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation.set_replication_additive" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the additive replication coefficients</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>coefficients: array of additive replication coefficients</li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">the -/+ basis is used throughout the library.
If you are used to the 0/1 basis, keep in mind that
the interaction series-expansion is different.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="FFPopSim.hivpopulation.set_resistance_additive">
<tt class="descname">set_resistance_additive</tt><big>(</big><em>coefficients</em><big>)</big><a class="headerlink" href="#FFPopSim.hivpopulation.set_resistance_additive" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the additive drug resistance coefficients</p>
<dl class="docutils">
<dt>Parameters:</dt>
<dd><ul class="first last simple">
<li>coefficients: array of additive drug resistance coefficients</li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">the -/+ basis is used throughout the library.
If you are used to the 0/1 basis, keep in mind that
the interaction series-expansion is different.</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="FFPopSim.hivpopulation.treatment">
<tt class="descname">treatment</tt><a class="headerlink" href="#FFPopSim.hivpopulation.treatment" title="Permalink to this definition">¶</a></dt>
<dd><p>Treatment weight (between 0 and 1)</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">this variable controls how important is either of the two phenotypic
traits, replication and resistance. Their contribution to fitness is
always linear (in this implementation).</p>
</div>
</dd></dl>

</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="haploid_highd.html"
                        title="previous chapter">haploid_highd Class Reference</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="hivgene.html"
                        title="next chapter">hivgene Class reference</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/hivpopulation.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="hivgene.html" title="hivgene Class reference"
             >next</a> |</li>
        <li class="right" >
          <a href="haploid_highd.html" title="haploid_highd Class Reference"
             >previous</a> |</li>
        <li><a href="index.html">FFPopSim 1.1 documentation</a> &raquo;</li>
          <li><a href="Contents.html" >Contents</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012, Fabio Zanini, Richard Neher.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>