<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>FFPopSim: haploid_highd Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    extensions: ["tex2jax.js", "TeX/AMSmath.js", "TeX/AMSsymbols.js"],
    jax: ["input/TeX","output/HTML-CSS"],
});
</script><script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">FFPopSim
   </div>
   <div id="projectbrief">Library for Population Genetics</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.2 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="inherits.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('classhaploid__highd.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Macros</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="#pro-methods">Protected Member Functions</a> &#124;
<a href="#pro-attribs">Protected Attributes</a> &#124;
<a href="classhaploid__highd-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">haploid_highd Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Population class for high-dimensional simulations.  
 <a href="classhaploid__highd.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="ffpopsim__highd_8h_source.html">ffpopsim_highd.h</a>&gt;</code></p>
<div class="dynheader">
Inheritance diagram for haploid_highd:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd__inherit__graph.png" border="0" usemap="#haploid__highd_inherit__map" alt="Inheritance graph"/></div>
<map name="haploid__highd_inherit__map" id="haploid__highd_inherit__map">
<area shape="rect" id="node3" href="classhivpopulation.html" title="HIV population with facultative drug treatment." alt="" coords="8,80,104,107"/></map>
<center><span class="legend">[<a target="top" href="graph_legend.html">legend</a>]</span></center></div>
<div class="dynheader">
Collaboration diagram for haploid_highd:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd__coll__graph.png" border="0" usemap="#haploid__highd_coll__map" alt="Collaboration graph"/></div>
<map name="haploid__highd_coll__map" id="haploid__highd_coll__map">
<area shape="rect" id="node2" href="structstat__t.html" title="Structure for short summary statistics." alt="" coords="5,5,59,32"/><area shape="rect" id="node4" href="classhypercube__highd.html" title="Hypercube class for high&#45;dimensional simulations." alt="" coords="84,5,204,32"/></map>
<center><span class="legend">[<a target="top" href="graph_legend.html">legend</a>]</span></center></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:ae58c4326bc251e8a86d3885f3297a25e"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ae58c4326bc251e8a86d3885f3297a25e">haploid_highd</a> (int <a class="el" href="classhaploid__highd.html#a246641ceeb4e3a8a7168426fc2bba212">L</a>=0, int rng_seed=0, int <a class="el" href="classhaploid__highd.html#af9bef51d5530a1807d79063fcaf81363">number_of_traits</a>=1)</td></tr>
<tr class="memdesc:ae58c4326bc251e8a86d3885f3297a25e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default constructor.  <a href="#ae58c4326bc251e8a86d3885f3297a25e"></a><br/></td></tr>
<tr class="separator:ae58c4326bc251e8a86d3885f3297a25e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4a8ea44da7eae169ae382d2840bd4185"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a4a8ea44da7eae169ae382d2840bd4185">~haploid_highd</a> ()</td></tr>
<tr class="memdesc:a4a8ea44da7eae169ae382d2840bd4185"><td class="mdescLeft">&#160;</td><td class="mdescRight">Destructor.  <a href="#a4a8ea44da7eae169ae382d2840bd4185"></a><br/></td></tr>
<tr class="separator:a4a8ea44da7eae169ae382d2840bd4185"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a246641ceeb4e3a8a7168426fc2bba212"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a246641ceeb4e3a8a7168426fc2bba212">L</a> ()</td></tr>
<tr class="separator:a246641ceeb4e3a8a7168426fc2bba212"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a303677b163c6b7edbf5f39a63ab84693"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a303677b163c6b7edbf5f39a63ab84693">get_number_of_loci</a> ()</td></tr>
<tr class="separator:a303677b163c6b7edbf5f39a63ab84693"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaa0a39abeeb60600158b8430739280ec"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#aaa0a39abeeb60600158b8430739280ec">N</a> ()</td></tr>
<tr class="separator:aaa0a39abeeb60600158b8430739280ec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5f2fe58daa26fce5a3cea94adf92fe97"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a5f2fe58daa26fce5a3cea94adf92fe97">get_population_size</a> ()</td></tr>
<tr class="separator:a5f2fe58daa26fce5a3cea94adf92fe97"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3a2a61046b0810e5b7b3794b49434aad"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a3a2a61046b0810e5b7b3794b49434aad">get_generation</a> ()</td></tr>
<tr class="separator:a3a2a61046b0810e5b7b3794b49434aad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6e915dc6199d83ba2e35b3195b5b03b9"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a6e915dc6199d83ba2e35b3195b5b03b9">get_number_of_clones</a> ()</td></tr>
<tr class="separator:a6e915dc6199d83ba2e35b3195b5b03b9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9e28c79bc0f9f68d705e0bf1643481f8"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a9e28c79bc0f9f68d705e0bf1643481f8">get_number_of_traits</a> ()</td></tr>
<tr class="separator:a9e28c79bc0f9f68d705e0bf1643481f8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2e29536be4ec4c75d2889e5a2a28642e"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a2e29536be4ec4c75d2889e5a2a28642e">get_participation_ratio</a> ()</td></tr>
<tr class="separator:a2e29536be4ec4c75d2889e5a2a28642e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3351949a49ad6702beb996b58755ce4e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a3351949a49ad6702beb996b58755ce4e">set_allele_frequencies</a> (double *freq, unsigned long <a class="el" href="classhaploid__highd.html#aaa0a39abeeb60600158b8430739280ec">N</a>)</td></tr>
<tr class="memdesc:a3351949a49ad6702beb996b58755ce4e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initialize population in linkage equilibrium.  <a href="#a3351949a49ad6702beb996b58755ce4e"></a><br/></td></tr>
<tr class="separator:a3351949a49ad6702beb996b58755ce4e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa0c34e02ed924603417fdb8109918f7e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#aa0c34e02ed924603417fdb8109918f7e">set_genotypes</a> (vector&lt; <a class="el" href="structgenotype__value__pair__t.html">genotype_value_pair_t</a> &gt; gt)</td></tr>
<tr class="memdesc:aa0c34e02ed924603417fdb8109918f7e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initialize the population with genotype counts.  <a href="#aa0c34e02ed924603417fdb8109918f7e"></a><br/></td></tr>
<tr class="separator:aa0c34e02ed924603417fdb8109918f7e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9418c16f238aacab3be2dbffb62246c4"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4">set_wildtype</a> (unsigned long <a class="el" href="classhaploid__highd.html#aaa0a39abeeb60600158b8430739280ec">N</a>)</td></tr>
<tr class="memdesc:a9418c16f238aacab3be2dbffb62246c4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initialize a wildtype population (00...0)  <a href="#a9418c16f238aacab3be2dbffb62246c4"></a><br/></td></tr>
<tr class="separator:a9418c16f238aacab3be2dbffb62246c4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a44235ff9f17f4e81f290611d7829818a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a44235ff9f17f4e81f290611d7829818a">add_genotype</a> (boost::dynamic_bitset&lt;&gt; genotype, int n=1)</td></tr>
<tr class="memdesc:a44235ff9f17f4e81f290611d7829818a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Add the genotype specified by a bitset to the current population in in n copies.  <a href="#a44235ff9f17f4e81f290611d7829818a"></a><br/></td></tr>
<tr class="separator:a44235ff9f17f4e81f290611d7829818a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac9da76630ffc3a1581f44f3e94004854"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ac9da76630ffc3a1581f44f3e94004854">add_trait_coefficient</a> (double value, vector&lt; int &gt; loci, int t=0)</td></tr>
<tr class="separator:ac9da76630ffc3a1581f44f3e94004854"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0a5523161a38427a7bdcc18136751817"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a0a5523161a38427a7bdcc18136751817">clear_trait</a> (int t=0)</td></tr>
<tr class="separator:a0a5523161a38427a7bdcc18136751817"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acc3e232638ff352b52a4c8db9b729a23"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#acc3e232638ff352b52a4c8db9b729a23">clear_traits</a> ()</td></tr>
<tr class="separator:acc3e232638ff352b52a4c8db9b729a23"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae8d94a3616a3c1ffdb07844fbd0349aa"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ae8d94a3616a3c1ffdb07844fbd0349aa">set_random_trait_epistasis</a> (double epistasis_std, int traitnumber=0)</td></tr>
<tr class="separator:ae8d94a3616a3c1ffdb07844fbd0349aa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af785069be8a626717d8e705b407e76a6"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#af785069be8a626717d8e705b407e76a6">add_fitness_coefficient</a> (double value, vector&lt; int &gt; loci)</td></tr>
<tr class="separator:af785069be8a626717d8e705b407e76a6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a298bdd5103be21db9f72fb72685cd414"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a298bdd5103be21db9f72fb72685cd414">clear_fitness</a> ()</td></tr>
<tr class="separator:a298bdd5103be21db9f72fb72685cd414"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0a7af3cbfabcbf0b3b01e4bcffd87425"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a0a7af3cbfabcbf0b3b01e4bcffd87425">set_random_epistasis</a> (double epistasis_std)</td></tr>
<tr class="separator:a0a7af3cbfabcbf0b3b01e4bcffd87425"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af7fbfc38495246eb5ad91201c9a8f58a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a">evolve</a> (int gen=1)</td></tr>
<tr class="memdesc:af7fbfc38495246eb5ad91201c9a8f58a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evolve for some generations under the specified conditions.  <a href="#af7fbfc38495246eb5ad91201c9a8f58a"></a><br/></td></tr>
<tr class="separator:af7fbfc38495246eb5ad91201c9a8f58a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aae3666ef1b744f0a1918d5694ec5199b"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#aae3666ef1b744f0a1918d5694ec5199b">bottleneck</a> (int size_of_bottleneck)</td></tr>
<tr class="memdesc:aae3666ef1b744f0a1918d5694ec5199b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Cause a bottleneck in the population size.  <a href="#aae3666ef1b744f0a1918d5694ec5199b"></a><br/></td></tr>
<tr class="separator:aae3666ef1b744f0a1918d5694ec5199b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af40993ccd48558777ab65ab88bc590b3"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3">calc_stat</a> ()</td></tr>
<tr class="memdesc:af40993ccd48558777ab65ab88bc590b3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate trait and fitness statistics and allele frequences.  <a href="#af40993ccd48558777ab65ab88bc590b3"></a><br/></td></tr>
<tr class="separator:af40993ccd48558777ab65ab88bc590b3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a981292b7434288cbee87093576b61371"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a981292b7434288cbee87093576b61371">unique_clones</a> ()</td></tr>
<tr class="memdesc:a981292b7434288cbee87093576b61371"><td class="mdescLeft">&#160;</td><td class="mdescRight">Remove duplicate clones.  <a href="#a981292b7434288cbee87093576b61371"></a><br/></td></tr>
<tr class="separator:a981292b7434288cbee87093576b61371"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab64531c02f245f39d9446efb5fcabe01"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ab64531c02f245f39d9446efb5fcabe01">random_clone</a> ()</td></tr>
<tr class="memdesc:ab64531c02f245f39d9446efb5fcabe01"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get a random clone from the population.  <a href="#ab64531c02f245f39d9446efb5fcabe01"></a><br/></td></tr>
<tr class="separator:ab64531c02f245f39d9446efb5fcabe01"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aefe07878a3133687465c41804043779d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#aefe07878a3133687465c41804043779d">random_clones</a> (unsigned int n_o_individuals, vector&lt; int &gt; *<a class="el" href="classsample.html">sample</a>)</td></tr>
<tr class="memdesc:aefe07878a3133687465c41804043779d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sample random individuals from the population.  <a href="#aefe07878a3133687465c41804043779d"></a><br/></td></tr>
<tr class="separator:aefe07878a3133687465c41804043779d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2f6a793822b830dbb67fdb2d22c22194"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a2f6a793822b830dbb67fdb2d22c22194">get_genotype_string</a> (unsigned int i)</td></tr>
<tr class="separator:a2f6a793822b830dbb67fdb2d22c22194"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a686aa6ae9c4d31f96dd6bb3d83d11188"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a686aa6ae9c4d31f96dd6bb3d83d11188">distance_Hamming</a> (unsigned int clone1, unsigned int clone2, vector&lt; unsigned int * &gt; *chunks=NULL, unsigned int every=1)</td></tr>
<tr class="separator:a686aa6ae9c4d31f96dd6bb3d83d11188"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aab7c4bf82bda5140c246b7d55cd2906d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#aab7c4bf82bda5140c246b7d55cd2906d">distance_Hamming</a> (boost::dynamic_bitset&lt;&gt; gt1, boost::dynamic_bitset&lt;&gt; gt2, vector&lt; unsigned int * &gt; *chunks=NULL, unsigned int every=1)</td></tr>
<tr class="memdesc:aab7c4bf82bda5140c246b7d55cd2906d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate Hamming distance between two sequences.  <a href="#aab7c4bf82bda5140c246b7d55cd2906d"></a><br/></td></tr>
<tr class="separator:aab7c4bf82bda5140c246b7d55cd2906d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abf5d073bfa998684b94c0b3248d7bdad"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structstat__t.html">stat_t</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#abf5d073bfa998684b94c0b3248d7bdad">get_diversity_statistics</a> (unsigned int n_sample=1000)</td></tr>
<tr class="memdesc:abf5d073bfa998684b94c0b3248d7bdad"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate diversity in the current population (Hamming distance between pairs of sequences)  <a href="#abf5d073bfa998684b94c0b3248d7bdad"></a><br/></td></tr>
<tr class="separator:abf5d073bfa998684b94c0b3248d7bdad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af77958954fe6001b2640b70bcc3681a4"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structstat__t.html">stat_t</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#af77958954fe6001b2640b70bcc3681a4">get_divergence_statistics</a> (unsigned int n_sample=1000)</td></tr>
<tr class="memdesc:af77958954fe6001b2640b70bcc3681a4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate mean and variance of the divergence from the [00...0] bitset.  <a href="#af77958954fe6001b2640b70bcc3681a4"></a><br/></td></tr>
<tr class="separator:af77958954fe6001b2640b70bcc3681a4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0c1eb3defc6dbf00e414482ae39355e2"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a0c1eb3defc6dbf00e414482ae39355e2">get_allele_frequency</a> (int l)</td></tr>
<tr class="separator:a0c1eb3defc6dbf00e414482ae39355e2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a03dee5a44b1ca487f2e7dd72dea3aabd"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a03dee5a44b1ca487f2e7dd72dea3aabd">get_pair_frequency</a> (int locus1, int locus2)</td></tr>
<tr class="memdesc:a03dee5a44b1ca487f2e7dd72dea3aabd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the joint frequency of two alleles.  <a href="#a03dee5a44b1ca487f2e7dd72dea3aabd"></a><br/></td></tr>
<tr class="separator:a03dee5a44b1ca487f2e7dd72dea3aabd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a061072acad96faee44962a5800f78281"><td class="memItemLeft" align="right" valign="top">vector&lt; double &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a061072acad96faee44962a5800f78281">get_pair_frequencies</a> (vector&lt; vector&lt; int &gt; &gt; *loci)</td></tr>
<tr class="memdesc:a061072acad96faee44962a5800f78281"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the joint frequency of two alleles, for a vector of allele pairs.  <a href="#a061072acad96faee44962a5800f78281"></a><br/></td></tr>
<tr class="separator:a061072acad96faee44962a5800f78281"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad94f599d9a5c7e3a13afe96a0a240d73"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ad94f599d9a5c7e3a13afe96a0a240d73">get_chi</a> (int l)</td></tr>
<tr class="separator:ad94f599d9a5c7e3a13afe96a0a240d73"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae87baa777bd8550c0d80ca44397a35b7"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ae87baa777bd8550c0d80ca44397a35b7">get_chi2</a> (int locus1, int locus2)</td></tr>
<tr class="separator:ae87baa777bd8550c0d80ca44397a35b7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a99d356f0eac94d6710f54b8a7f303f4a"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a99d356f0eac94d6710f54b8a7f303f4a">get_LD</a> (int locus1, int locus2)</td></tr>
<tr class="separator:a99d356f0eac94d6710f54b8a7f303f4a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad4c1f8246b0c6ae14745d9634ff04ee6"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ad4c1f8246b0c6ae14745d9634ff04ee6">get_moment</a> (int locus1, int locus2)</td></tr>
<tr class="separator:ad4c1f8246b0c6ae14745d9634ff04ee6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac6029079cbb7ff559ed75b581ff8deef"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ac6029079cbb7ff559ed75b581ff8deef">get_fitness</a> (int n)</td></tr>
<tr class="separator:ac6029079cbb7ff559ed75b581ff8deef"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8351db0129fbfb98170f0b60565c9dc3"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a8351db0129fbfb98170f0b60565c9dc3">get_clone_size</a> (int n)</td></tr>
<tr class="separator:a8351db0129fbfb98170f0b60565c9dc3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1da3a90a51c42ccccbd9c47a93817052"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a1da3a90a51c42ccccbd9c47a93817052">get_trait</a> (int n, int t=0)</td></tr>
<tr class="separator:a1da3a90a51c42ccccbd9c47a93817052"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4271fb90101fdea28032029efffbd395"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structstat__t.html">stat_t</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a4271fb90101fdea28032029efffbd395">get_fitness_statistics</a> ()</td></tr>
<tr class="separator:a4271fb90101fdea28032029efffbd395"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3554acf03e30ec4df69de98c29e1b2a8"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structstat__t.html">stat_t</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a3554acf03e30ec4df69de98c29e1b2a8">get_trait_statistics</a> (int t=0)</td></tr>
<tr class="separator:a3554acf03e30ec4df69de98c29e1b2a8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5c6adc6e23625ec7eda23253074b564c"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a5c6adc6e23625ec7eda23253074b564c">get_trait_covariance</a> (int t1, int t2)</td></tr>
<tr class="separator:a5c6adc6e23625ec7eda23253074b564c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7d00bee54a9dcddcec55d873545047cb"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a7d00bee54a9dcddcec55d873545047cb">get_max_fitness</a> ()</td></tr>
<tr class="separator:a7d00bee54a9dcddcec55d873545047cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad5eee5bc7f757e20fb3119bb01c59b1e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ad5eee5bc7f757e20fb3119bb01c59b1e">update_traits</a> ()</td></tr>
<tr class="memdesc:ad5eee5bc7f757e20fb3119bb01c59b1e"><td class="mdescLeft">&#160;</td><td class="mdescRight">For each clone, recalculate its traits.  <a href="#ad5eee5bc7f757e20fb3119bb01c59b1e"></a><br/></td></tr>
<tr class="separator:ad5eee5bc7f757e20fb3119bb01c59b1e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9b8ada9362e3108ee8994634231541be"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a9b8ada9362e3108ee8994634231541be">update_fitness</a> ()</td></tr>
<tr class="memdesc:a9b8ada9362e3108ee8994634231541be"><td class="mdescLeft">&#160;</td><td class="mdescRight">For each clone, update fitness assuming traits are already up to date.  <a href="#a9b8ada9362e3108ee8994634231541be"></a><br/></td></tr>
<tr class="separator:a9b8ada9362e3108ee8994634231541be"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a234a418eff9319a2dfb6371b73220b56"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a234a418eff9319a2dfb6371b73220b56">get_divergence_histogram</a> (gsl_histogram **hist, unsigned int bins=10, vector&lt; unsigned int * &gt; *chunks=NULL, unsigned int every=1, unsigned int n_sample=1000)</td></tr>
<tr class="memdesc:a234a418eff9319a2dfb6371b73220b56"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get histogram of divergence from the [00...0] bitset.  <a href="#a234a418eff9319a2dfb6371b73220b56"></a><br/></td></tr>
<tr class="separator:a234a418eff9319a2dfb6371b73220b56"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a25a490ade017cff4cccb88ced04261ce"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a25a490ade017cff4cccb88ced04261ce">get_diversity_histogram</a> (gsl_histogram **hist, unsigned int bins=10, vector&lt; unsigned int * &gt; *chunks=NULL, unsigned int every=1, unsigned int n_sample=1000)</td></tr>
<tr class="memdesc:a25a490ade017cff4cccb88ced04261ce"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get histogram of diversity in the population (mutual Hamming distance)  <a href="#a25a490ade017cff4cccb88ced04261ce"></a><br/></td></tr>
<tr class="separator:a25a490ade017cff4cccb88ced04261ce"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2dae66421820a2248db9be1771ce9e62"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a2dae66421820a2248db9be1771ce9e62">get_fitness_histogram</a> (gsl_histogram **hist, unsigned int bins=10, unsigned int n_sample=1000)</td></tr>
<tr class="memdesc:a2dae66421820a2248db9be1771ce9e62"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate histogram of fitness from traits.  <a href="#a2dae66421820a2248db9be1771ce9e62"></a><br/></td></tr>
<tr class="separator:a2dae66421820a2248db9be1771ce9e62"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae25ec86176f438db2dc0194f04ecfc42"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ae25ec86176f438db2dc0194f04ecfc42">print_allele_frequencies</a> (ostream &amp;out)</td></tr>
<tr class="memdesc:ae25ec86176f438db2dc0194f04ecfc42"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print all allele frequencies into a stream provided.  <a href="#ae25ec86176f438db2dc0194f04ecfc42"></a><br/></td></tr>
<tr class="separator:ae25ec86176f438db2dc0194f04ecfc42"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a60617675023cab02a599a1ea7f9605c6"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a60617675023cab02a599a1ea7f9605c6">read_ms_sample</a> (istream &amp;gts, int skip_locus, int multiplicity)</td></tr>
<tr class="memdesc:a60617675023cab02a599a1ea7f9605c6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read the output of Hudson's ms and use it to initialize the genotype distribution.  <a href="#a60617675023cab02a599a1ea7f9605c6"></a><br/></td></tr>
<tr class="separator:a60617675023cab02a599a1ea7f9605c6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0e08d1847e521402cf0e2676d6ed38b9"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9">read_ms_sample_sparse</a> (istream &amp;gts, int skip_locus, int multiplicity, int distance)</td></tr>
<tr class="memdesc:a0e08d1847e521402cf0e2676d6ed38b9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read the output of Hudson's ms and use it to initialize the genotype distribution.  <a href="#a0e08d1847e521402cf0e2676d6ed38b9"></a><br/></td></tr>
<tr class="separator:a0e08d1847e521402cf0e2676d6ed38b9"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a3faa49df0828300313fb979e29b53bd8"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classhypercube__highd.html">hypercube_highd</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a3faa49df0828300313fb979e29b53bd8">trait</a></td></tr>
<tr class="separator:a3faa49df0828300313fb979e29b53bd8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7b2a100e6109b0190f3435b08097c327"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="structclone__t.html">clone_t</a> &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a7b2a100e6109b0190f3435b08097c327">current_pop</a></td></tr>
<tr class="separator:a7b2a100e6109b0190f3435b08097c327"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad0f15c886bb03789063f5587e080166e"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="structclone__t.html">clone_t</a> &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ad0f15c886bb03789063f5587e080166e">new_pop</a></td></tr>
<tr class="separator:ad0f15c886bb03789063f5587e080166e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa496b4cc82e39329cc6d18b1432cd515"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#aa496b4cc82e39329cc6d18b1432cd515">carrying_capacity</a></td></tr>
<tr class="separator:aa496b4cc82e39329cc6d18b1432cd515"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad55b1105f6869684863c7e4f34023827"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ad55b1105f6869684863c7e4f34023827">mutation_rate</a></td></tr>
<tr class="separator:ad55b1105f6869684863c7e4f34023827"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4dc7a29976c59a5c4755d12ade075069"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a4dc7a29976c59a5c4755d12ade075069">outcrossing_rate</a></td></tr>
<tr class="separator:a4dc7a29976c59a5c4755d12ade075069"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a023aab1a8360b632e471bdc3a4ec37d1"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a023aab1a8360b632e471bdc3a4ec37d1">crossover_rate</a></td></tr>
<tr class="separator:a023aab1a8360b632e471bdc3a4ec37d1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afa9a75efb02352d951e633f31ac1bc59"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#afa9a75efb02352d951e633f31ac1bc59">recombination_model</a></td></tr>
<tr class="separator:afa9a75efb02352d951e633f31ac1bc59"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abc8e66940a4dd00a25388ec322cf7a49"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#abc8e66940a4dd00a25388ec322cf7a49">circular</a></td></tr>
<tr class="separator:abc8e66940a4dd00a25388ec322cf7a49"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-methods"></a>
Protected Member Functions</h2></td></tr>
<tr class="memitem:a643f680f350c9060df4b01cd856cc50e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a643f680f350c9060df4b01cd856cc50e">get_random_seed</a> ()</td></tr>
<tr class="memdesc:a643f680f350c9060df4b01cd856cc50e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get a random seed from /dev/urandom.  <a href="#a643f680f350c9060df4b01cd856cc50e"></a><br/></td></tr>
<tr class="separator:a643f680f350c9060df4b01cd856cc50e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab3e442c79f83e603bc30892e2a36e329"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ab3e442c79f83e603bc30892e2a36e329">produce_random_sample</a> (int size=1000)</td></tr>
<tr class="memdesc:ab3e442c79f83e603bc30892e2a36e329"><td class="mdescLeft">&#160;</td><td class="mdescRight">Produce and store a random sample of the population for stochastic processes.  <a href="#ab3e442c79f83e603bc30892e2a36e329"></a><br/></td></tr>
<tr class="separator:ab3e442c79f83e603bc30892e2a36e329"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af173582ee41aecd03d8f97aa234cf879"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#af173582ee41aecd03d8f97aa234cf879">mutate</a> ()</td></tr>
<tr class="memdesc:af173582ee41aecd03d8f97aa234cf879"><td class="mdescLeft">&#160;</td><td class="mdescRight">Mutate random clones at random loci (all loci)  <a href="#af173582ee41aecd03d8f97aa234cf879"></a><br/></td></tr>
<tr class="separator:af173582ee41aecd03d8f97aa234cf879"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aec2c31155f262a9c1cf21262eea2d03a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#aec2c31155f262a9c1cf21262eea2d03a">select_gametes</a> ()</td></tr>
<tr class="memdesc:aec2c31155f262a9c1cf21262eea2d03a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generate offspring according to fitness (selection) and segregate some for sexual mating.  <a href="#aec2c31155f262a9c1cf21262eea2d03a"></a><br/></td></tr>
<tr class="separator:aec2c31155f262a9c1cf21262eea2d03a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae64df329db764e98e77398a775a43600"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ae64df329db764e98e77398a775a43600">relaxation_value</a> ()</td></tr>
<tr class="memdesc:ae64df329db764e98e77398a775a43600"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the log of the exp-average fitness plus relaxation term.  <a href="#ae64df329db764e98e77398a775a43600"></a><br/></td></tr>
<tr class="separator:ae64df329db764e98e77398a775a43600"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0aaa82167545768f814c7e28868eee77"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a0aaa82167545768f814c7e28868eee77">get_logmean_expfitness</a> ()</td></tr>
<tr class="memdesc:a0aaa82167545768f814c7e28868eee77"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the population exp-average of fitness, used for keeping the population size fixed.  <a href="#a0aaa82167545768f814c7e28868eee77"></a><br/></td></tr>
<tr class="separator:a0aaa82167545768f814c7e28868eee77"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a943354008a9931e4223f952f4790efd6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a943354008a9931e4223f952f4790efd6">flip_single_locus</a> (unsigned int clonenum, int locus)</td></tr>
<tr class="memdesc:a943354008a9931e4223f952f4790efd6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flip a spin (locus) in individual.  <a href="#a943354008a9931e4223f952f4790efd6"></a><br/></td></tr>
<tr class="separator:a943354008a9931e4223f952f4790efd6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a710a61fe8975aaaf4d148b2da20985ae"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a710a61fe8975aaaf4d148b2da20985ae">flip_single_locus</a> (int locus)</td></tr>
<tr class="memdesc:a710a61fe8975aaaf4d148b2da20985ae"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flip a spin at a specific locus in random individual.  <a href="#a710a61fe8975aaaf4d148b2da20985ae"></a><br/></td></tr>
<tr class="separator:a710a61fe8975aaaf4d148b2da20985ae"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6c8674a3f1dcb55ea680b2fd7dbf655c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a6c8674a3f1dcb55ea680b2fd7dbf655c">shuffle_genotypes</a> ()</td></tr>
<tr class="separator:a6c8674a3f1dcb55ea680b2fd7dbf655c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6fa4ffe295eef314642dec0208f81b7d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a6fa4ffe295eef314642dec0208f81b7d">swap_populations</a> ()</td></tr>
<tr class="memdesc:a6fa4ffe295eef314642dec0208f81b7d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Make the the temporary population the current one.  <a href="#a6fa4ffe295eef314642dec0208f81b7d"></a><br/></td></tr>
<tr class="separator:a6fa4ffe295eef314642dec0208f81b7d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab35ed279920654c5010561b692da200b"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ab35ed279920654c5010561b692da200b">new_generation</a> ()</td></tr>
<tr class="separator:ab35ed279920654c5010561b692da200b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af9d9998d0510c5fde8d3a422f900d731"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#af9d9998d0510c5fde8d3a422f900d731">partition_cumulative</a> (vector&lt; unsigned int &gt; &amp;partition_cum)</td></tr>
<tr class="memdesc:af9d9998d0510c5fde8d3a422f900d731"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the cumulative partition of sequences into clones.  <a href="#af9d9998d0510c5fde8d3a422f900d731"></a><br/></td></tr>
<tr class="separator:af9d9998d0510c5fde8d3a422f900d731"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a93cb185958cc78ce66c909d8df532a57"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a93cb185958cc78ce66c909d8df532a57">calc_allele_freqs</a> ()</td></tr>
<tr class="memdesc:a93cb185958cc78ce66c909d8df532a57"><td class="mdescLeft">&#160;</td><td class="mdescRight">calculate and store allele frequencies  <a href="#a93cb185958cc78ce66c909d8df532a57"></a><br/></td></tr>
<tr class="separator:a93cb185958cc78ce66c909d8df532a57"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac79972f0e845b52b3191a27ddae309a8"><td class="memItemLeft" align="right" valign="top">boost::dynamic_bitset&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ac79972f0e845b52b3191a27ddae309a8">reassortment_pattern</a> ()</td></tr>
<tr class="memdesc:ac79972f0e845b52b3191a27ddae309a8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Produce a random reassortement pattern.  <a href="#ac79972f0e845b52b3191a27ddae309a8"></a><br/></td></tr>
<tr class="separator:ac79972f0e845b52b3191a27ddae309a8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a33a62c8a5c57d36e071099906131eb31"><td class="memItemLeft" align="right" valign="top">boost::dynamic_bitset&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a33a62c8a5c57d36e071099906131eb31">crossover_pattern</a> ()</td></tr>
<tr class="memdesc:a33a62c8a5c57d36e071099906131eb31"><td class="mdescLeft">&#160;</td><td class="mdescRight">Choose a number of crossover points and produce a crossover pattern.  <a href="#a33a62c8a5c57d36e071099906131eb31"></a><br/></td></tr>
<tr class="separator:a33a62c8a5c57d36e071099906131eb31"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab08783cb783ca66ff2c87227dbd74314"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ab08783cb783ca66ff2c87227dbd74314">add_recombinants</a> ()</td></tr>
<tr class="memdesc:ab08783cb783ca66ff2c87227dbd74314"><td class="mdescLeft">&#160;</td><td class="mdescRight">Pair and mate sexual gametes.  <a href="#ab08783cb783ca66ff2c87227dbd74314"></a><br/></td></tr>
<tr class="separator:ab08783cb783ca66ff2c87227dbd74314"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afb3d9e695f16b8412902757f18cfb63e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#afb3d9e695f16b8412902757f18cfb63e">recombine</a> (int parent1, int parent2)</td></tr>
<tr class="memdesc:afb3d9e695f16b8412902757f18cfb63e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Recombine two genotypes parent1 and parent2 to produce two new genotypes.  <a href="#afb3d9e695f16b8412902757f18cfb63e"></a><br/></td></tr>
<tr class="separator:afb3d9e695f16b8412902757f18cfb63e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae9c993847dc49cd24811df7ef1fb9a5e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ae9c993847dc49cd24811df7ef1fb9a5e">recombine_crossover</a> (int parent1, int parent2, int ng)</td></tr>
<tr class="separator:ae9c993847dc49cd24811df7ef1fb9a5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6dd67288b3fc216d2b83d7e4861966af"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a6dd67288b3fc216d2b83d7e4861966af">calc_fitness_stat</a> ()</td></tr>
<tr class="memdesc:a6dd67288b3fc216d2b83d7e4861966af"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate and store fitness population statistics.  <a href="#a6dd67288b3fc216d2b83d7e4861966af"></a><br/></td></tr>
<tr class="separator:a6dd67288b3fc216d2b83d7e4861966af"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afb3de45e613ffe03eadcc10f81c13448"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#afb3de45e613ffe03eadcc10f81c13448">calc_trait_stat</a> ()</td></tr>
<tr class="memdesc:afb3de45e613ffe03eadcc10f81c13448"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate and store trait population statistics and covariances.  <a href="#afb3de45e613ffe03eadcc10f81c13448"></a><br/></td></tr>
<tr class="separator:afb3de45e613ffe03eadcc10f81c13448"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acf40ae60f1441f5a8bb67a5a84444c01"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01">calc_individual_traits</a> (<a class="el" href="structclone__t.html">clone_t</a> *tempgt)</td></tr>
<tr class="memdesc:acf40ae60f1441f5a8bb67a5a84444c01"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate traits of the chosen clone.  <a href="#acf40ae60f1441f5a8bb67a5a84444c01"></a><br/></td></tr>
<tr class="separator:acf40ae60f1441f5a8bb67a5a84444c01"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1ee15d862374062377099130afc62c20"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a1ee15d862374062377099130afc62c20">calc_individual_fitness</a> (<a class="el" href="structclone__t.html">clone_t</a> *tempgt)</td></tr>
<tr class="memdesc:a1ee15d862374062377099130afc62c20"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate fitness of a particular clone.  <a href="#a1ee15d862374062377099130afc62c20"></a><br/></td></tr>
<tr class="separator:a1ee15d862374062377099130afc62c20"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7f51d6a4f5791961d1b262e15735f820"><td class="memItemLeft" align="right" valign="top">virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820">calc_individual_fitness_from_traits</a> (<a class="el" href="structclone__t.html">clone_t</a> *tempgt)</td></tr>
<tr class="memdesc:a7f51d6a4f5791961d1b262e15735f820"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate fitness from traits of the chosen clone.  <a href="#a7f51d6a4f5791961d1b262e15735f820"></a><br/></td></tr>
<tr class="separator:a7f51d6a4f5791961d1b262e15735f820"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0d929c061bbf1c2fd4d82f6bb2f2f998"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998">check_individual_maximal_fitness</a> (<a class="el" href="structclone__t.html">clone_t</a> *tempgt)</td></tr>
<tr class="separator:a0d929c061bbf1c2fd4d82f6bb2f2f998"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac4d5cd6a0e86ae30257f41c1e5b889c5"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ac4d5cd6a0e86ae30257f41c1e5b889c5">get_trait_difference</a> (<a class="el" href="structclone__t.html">clone_t</a> *tempgt1, <a class="el" href="structclone__t.html">clone_t</a> *tempgt2, vector&lt; int &gt; &amp;diffpos, int traitnum)</td></tr>
<tr class="memdesc:ac4d5cd6a0e86ae30257f41c1e5b889c5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate trait difference between two clones.  <a href="#ac4d5cd6a0e86ae30257f41c1e5b889c5"></a><br/></td></tr>
<tr class="separator:ac4d5cd6a0e86ae30257f41c1e5b889c5"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-attribs"></a>
Protected Attributes</h2></td></tr>
<tr class="memitem:afe01d72044a8d81a32694287aa68ef09"><td class="memItemLeft" align="right" valign="top">gsl_rng *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#afe01d72044a8d81a32694287aa68ef09">evo_generator</a></td></tr>
<tr class="separator:afe01d72044a8d81a32694287aa68ef09"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9395ef40e6478fbc361c82193377346e"><td class="memItemLeft" align="right" valign="top">gsl_rng *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a9395ef40e6478fbc361c82193377346e">label_generator</a></td></tr>
<tr class="separator:a9395ef40e6478fbc361c82193377346e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1d7ee5c309f75b8b5d97f087569c532d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a1d7ee5c309f75b8b5d97f087569c532d">seed</a></td></tr>
<tr class="separator:a1d7ee5c309f75b8b5d97f087569c532d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1c647247fb891a314caf21b58f5a1e0d"><td class="memItemLeft" align="right" valign="top">vector&lt; int &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a1c647247fb891a314caf21b58f5a1e0d">random_sample</a></td></tr>
<tr class="separator:a1c647247fb891a314caf21b58f5a1e0d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad1a76259095c3119e33fcc63abd52227"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ad1a76259095c3119e33fcc63abd52227">number_of_loci</a></td></tr>
<tr class="separator:ad1a76259095c3119e33fcc63abd52227"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a13f8881a0a81b4cc1ac3fe23e95c27fd"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a13f8881a0a81b4cc1ac3fe23e95c27fd">population_size</a></td></tr>
<tr class="separator:a13f8881a0a81b4cc1ac3fe23e95c27fd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af9bef51d5530a1807d79063fcaf81363"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#af9bef51d5530a1807d79063fcaf81363">number_of_traits</a></td></tr>
<tr class="separator:af9bef51d5530a1807d79063fcaf81363"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad364734ea945abb84658303eefa0d817"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ad364734ea945abb84658303eefa0d817">scratch</a></td></tr>
<tr class="separator:ad364734ea945abb84658303eefa0d817"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a771e8b659f399927ed612ba2f01c7ef8"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a771e8b659f399927ed612ba2f01c7ef8">generation</a></td></tr>
<tr class="separator:a771e8b659f399927ed612ba2f01c7ef8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2929fa585f451eff209d5766e280d0b6"><td class="memItemLeft" align="right" valign="top">double *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a2929fa585f451eff209d5766e280d0b6">allele_frequencies</a></td></tr>
<tr class="separator:a2929fa585f451eff209d5766e280d0b6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abcade2de00cd6efee795651d46eede3a"><td class="memItemLeft" align="right" valign="top">double *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#abcade2de00cd6efee795651d46eede3a">gamete_allele_frequencies</a></td></tr>
<tr class="separator:abcade2de00cd6efee795651d46eede3a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1f25e3ce87377569a3b14c92bb72fc55"><td class="memItemLeft" align="right" valign="top">double *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a1f25e3ce87377569a3b14c92bb72fc55">chi1</a></td></tr>
<tr class="separator:a1f25e3ce87377569a3b14c92bb72fc55"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0047d21da6fe6fa9c1ce94b7b53ae37a"><td class="memItemLeft" align="right" valign="top">double **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a0047d21da6fe6fa9c1ce94b7b53ae37a">chi2</a></td></tr>
<tr class="separator:a0047d21da6fe6fa9c1ce94b7b53ae37a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac0c344e3d804d8f788ea527167a9bddc"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ac0c344e3d804d8f788ea527167a9bddc">outcrossing_rate_effective</a></td></tr>
<tr class="separator:ac0c344e3d804d8f788ea527167a9bddc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8a44ffb375d6763ebdef28d9e5121d8d"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a8a44ffb375d6763ebdef28d9e5121d8d">genome</a></td></tr>
<tr class="separator:a8a44ffb375d6763ebdef28d9e5121d8d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a495c975322ea03d01417be29e951a1f0"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a495c975322ea03d01417be29e951a1f0">crossovers</a></td></tr>
<tr class="separator:a495c975322ea03d01417be29e951a1f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a720e92ddf95f0aaf1e5136b60b53caad"><td class="memItemLeft" align="right" valign="top">vector&lt; int &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a720e92ddf95f0aaf1e5136b60b53caad">sex_gametes</a></td></tr>
<tr class="separator:a720e92ddf95f0aaf1e5136b60b53caad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1a57b48e7395ef197c38c46329af820a"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a1a57b48e7395ef197c38c46329af820a">fitness_max</a></td></tr>
<tr class="separator:a1a57b48e7395ef197c38c46329af820a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae57c154a248e0cfe9be446a04d1765d5"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#ae57c154a248e0cfe9be446a04d1765d5">participation_ratio</a></td></tr>
<tr class="separator:ae57c154a248e0cfe9be446a04d1765d5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa3cd0c02607b41b378f7fa488d5b5b46"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structstat__t.html">stat_t</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#aa3cd0c02607b41b378f7fa488d5b5b46">fitness_stat</a></td></tr>
<tr class="separator:aa3cd0c02607b41b378f7fa488d5b5b46"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a97fe03c6b3b3034410d728f17fa2716b"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structstat__t.html">stat_t</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a97fe03c6b3b3034410d728f17fa2716b">trait_stat</a></td></tr>
<tr class="separator:a97fe03c6b3b3034410d728f17fa2716b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a870cd9eb97be85b79de12baba7231aa7"><td class="memItemLeft" align="right" valign="top">double **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classhaploid__highd.html#a870cd9eb97be85b79de12baba7231aa7">trait_covariance</a></td></tr>
<tr class="separator:a870cd9eb97be85b79de12baba7231aa7"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Population class for high-dimensional simulations. </p>
<p>This class is the main object storing the state of and enabling the manipulation of populations with long genomes ( \(L\)) larger than 20.</p>
<p>Both asexual and sexual populations can be simulated. Since asexual populations under selection are often structured as a small list of large clones, the class stores the clones and their sizes instead of the individuals.</p>
<p>Class methods give access to a variety of information on the population, including:</p>
<ul>
<li>the fitness distribution;</li>
<li>summary statistics of fitness and other phenotypic trits;</li>
<li>genetic structure (linkage disequilibrium, allele frequencies, number of clones). </li>
</ul>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="ae58c4326bc251e8a86d3885f3297a25e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">haploid_highd::haploid_highd </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>L_in</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>rng_seed</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n_o_traits</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Default constructor. </p>
<p>The population objects are initialized, but neither the random generator is initialized nor the memory for traits, gametes etc. is allocated. Please call set_up() before using the class.</p>
<p>Note: The sequence is assumed to be linear (not circular). You can change this by hand if you wish so. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ae58c4326bc251e8a86d3885f3297a25e_cgraph.png" border="0" usemap="#classhaploid__highd_ae58c4326bc251e8a86d3885f3297a25e_cgraph" alt=""/></div>
<map name="classhaploid__highd_ae58c4326bc251e8a86d3885f3297a25e_cgraph" id="classhaploid__highd_ae58c4326bc251e8a86d3885f3297a25e_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a643f680f350c9060df4b01cd856cc50e" title="Get a random seed from /dev/urandom." alt="" coords="206,5,333,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a4a8ea44da7eae169ae382d2840bd4185"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">haploid_highd::~haploid_highd </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Destructor. </p>
<p>Memory is released here. </p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="af785069be8a626717d8e705b407e76a6"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::add_fitness_coefficient </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>value</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; int &gt;&#160;</td>
          <td class="paramname"><em>loci</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a44235ff9f17f4e81f290611d7829818a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::add_genotype </td>
          <td>(</td>
          <td class="paramtype">boost::dynamic_bitset&lt;&gt;&#160;</td>
          <td class="paramname"><em>genotype</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Add the genotype specified by a bitset to the current population in in n copies. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">genotype</td><td>genotype being added </td></tr>
    <tr><td class="paramname">n</td><td>number of copies of that genotype</td></tr>
  </table>
  </dd>
</dl>
<p>Note: this function also calculates the traits and fitness of the new individual. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a44235ff9f17f4e81f290611d7829818a_cgraph.png" border="0" usemap="#classhaploid__highd_a44235ff9f17f4e81f290611d7829818a_cgraph" alt=""/></div>
<map name="classhaploid__highd_a44235ff9f17f4e81f290611d7829818a_cgraph" id="classhaploid__highd_a44235ff9f17f4e81f290611d7829818a_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="209,5,343,46"/><area shape="rect" id="node7" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="209,69,343,125"/><area shape="rect" id="node9" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="185,149,367,190"/><area shape="rect" id="node5" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="417,12,594,39"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a44235ff9f17f4e81f290611d7829818a_icgraph.png" border="0" usemap="#classhaploid__highd_a44235ff9f17f4e81f290611d7829818a_icgraph" alt=""/></div>
<map name="classhaploid__highd_a44235ff9f17f4e81f290611d7829818a_icgraph" id="classhaploid__highd_a44235ff9f17f4e81f290611d7829818a_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a3351949a49ad6702beb996b58755ce4e" title="Initialize population in linkage equilibrium." alt="" coords="185,5,317,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#aa0c34e02ed924603417fdb8109918f7e" title="Initialize the population with genotype counts." alt="" coords="187,70,314,111"/><area shape="rect" id="node7" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4" title="Initialize a wildtype population (00...0)" alt="" coords="187,187,314,229"/><area shape="rect" id="node9" href="classhaploid__highd.html#a60617675023cab02a599a1ea7f9605c6" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="387,129,522,170"/><area shape="rect" id="node11" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="383,246,526,287"/><area shape="rect" id="node13" href="classhivpopulation.html#a299863ae882da177d548edc26d9a2cb8" title="Construct a HIV population with certain parameters." alt="" coords="365,195,544,221"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ab08783cb783ca66ff2c87227dbd74314"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::add_recombinants </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Pair and mate sexual gametes. </p>
<dl class="section return"><dt>Returns</dt><dd>zero if successful, nonzero otherwise</dd></dl>
<p>Using the previously produced list of sex_gametes, pair them at random and mate </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ab08783cb783ca66ff2c87227dbd74314_cgraph.png" border="0" usemap="#classhaploid__highd_ab08783cb783ca66ff2c87227dbd74314_cgraph" alt=""/></div>
<map name="classhaploid__highd_ab08783cb783ca66ff2c87227dbd74314_cgraph" id="classhaploid__highd_ab08783cb783ca66ff2c87227dbd74314_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#afb3d9e695f16b8412902757f18cfb63e" title="Recombine two genotypes parent1 and parent2 to produce two new genotypes." alt="" coords="185,143,354,169"/><area shape="rect" id="node5" href="classhaploid__highd.html#ac79972f0e845b52b3191a27ddae309a8" title="Produce a random reassortement pattern." alt="" coords="403,5,589,46"/><area shape="rect" id="node7" href="classhaploid__highd.html#a33a62c8a5c57d36e071099906131eb31" title="Choose a number of crossover points and produce a crossover pattern." alt="" coords="414,70,578,111"/><area shape="rect" id="node9" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="429,135,563,177"/><area shape="rect" id="node13" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="429,200,563,256"/><area shape="rect" id="node15" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="405,279,587,321"/><area shape="rect" id="node11" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="638,143,815,169"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ab08783cb783ca66ff2c87227dbd74314_icgraph.png" border="0" usemap="#classhaploid__highd_ab08783cb783ca66ff2c87227dbd74314_icgraph" alt=""/></div>
<map name="classhaploid__highd_ab08783cb783ca66ff2c87227dbd74314_icgraph" id="classhaploid__highd_ab08783cb783ca66ff2c87227dbd74314_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="185,12,327,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ac9da76630ffc3a1581f44f3e94004854"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::add_trait_coefficient </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>value</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; int &gt;&#160;</td>
          <td class="paramname"><em>loci</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>t</em> = <code>0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ac9da76630ffc3a1581f44f3e94004854_icgraph.png" border="0" usemap="#classhaploid__highd_ac9da76630ffc3a1581f44f3e94004854_icgraph" alt=""/></div>
<map name="classhaploid__highd_ac9da76630ffc3a1581f44f3e94004854_icgraph" id="classhaploid__highd_ac9da76630ffc3a1581f44f3e94004854_icgraph">
<area shape="rect" id="node3" href="classhivpopulation.html#ad1caa476f2d93020d52d0ce46f81a529" title="hivpopulation::read\l_replication_coefficients" alt="" coords="185,5,345,46"/><area shape="rect" id="node5" href="classhivpopulation.html#a6afaadd90cf25378476a13a2bef80875" title="hivpopulation::read\l_resistance_coefficients" alt="" coords="185,70,346,111"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aae3666ef1b744f0a1918d5694ec5199b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::bottleneck </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>size_of_bottleneck</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Cause a bottleneck in the population size. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">size_of_bottleneck</td><td>number of individuals to leave alive (approximate)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>zero if successful, error codes otherwise</dd></dl>
<p>The bottleneck is performed as follows. Each clone is traded in for a smaller one with size given by a Poisson random number around the new population size times the clone frequency. This function should therefore almost conserve genotype frequencies, except for rare genotypes that are lost.</p>
<p>TODO: this function should accept a gsl random distribution as optional argument for choosing how sharp the bottleneck should be (i.e., how large fluctuations around the expected frequency may be). However, this requires function pointers or templates or lambda functions, and might be a nightmare to code. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_aae3666ef1b744f0a1918d5694ec5199b_cgraph.png" border="0" usemap="#classhaploid__highd_aae3666ef1b744f0a1918d5694ec5199b_cgraph" alt=""/></div>
<map name="classhaploid__highd_aae3666ef1b744f0a1918d5694ec5199b_cgraph" id="classhaploid__highd_aae3666ef1b744f0a1918d5694ec5199b_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a6fa4ffe295eef314642dec0208f81b7d" title="Make the the temporary population the current one." alt="" coords="224,5,363,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a93cb185958cc78ce66c909d8df532a57"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::calc_allele_freqs </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>calculate and store allele frequencies </p>
<p>Note: the allele frequencies are available in the allele_frequencies attribute. </p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a93cb185958cc78ce66c909d8df532a57_icgraph.png" border="0" usemap="#classhaploid__highd_a93cb185958cc78ce66c909d8df532a57_icgraph" alt=""/></div>
<map name="classhaploid__highd_a93cb185958cc78ce66c909d8df532a57_icgraph" id="classhaploid__highd_a93cb185958cc78ce66c909d8df532a57_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="187,111,349,137"/><area shape="rect" id="node5" href="classhaploid__highd.html#a3351949a49ad6702beb996b58755ce4e" title="Initialize population in linkage equilibrium." alt="" coords="399,5,531,46"/><area shape="rect" id="node7" href="classhaploid__highd.html#aa0c34e02ed924603417fdb8109918f7e" title="Initialize the population with genotype counts." alt="" coords="402,70,529,111"/><area shape="rect" id="node9" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4" title="Initialize a wildtype population (00...0)" alt="" coords="402,135,529,177"/><area shape="rect" id="node17" href="classhaploid__highd.html#ae25ec86176f438db2dc0194f04ecfc42" title="Print all allele frequencies into a stream provided." alt="" coords="398,201,533,242"/><area shape="rect" id="node11" href="classhaploid__highd.html#a60617675023cab02a599a1ea7f9605c6" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="603,70,738,111"/><area shape="rect" id="node13" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="599,135,742,177"/><area shape="rect" id="node15" href="classhivpopulation.html#a299863ae882da177d548edc26d9a2cb8" title="Construct a HIV population with certain parameters." alt="" coords="581,201,760,228"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a6dd67288b3fc216d2b83d7e4861966af"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::calc_fitness_stat </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Calculate and store fitness population statistics. </p>
<p>The fitness statistics are stored in fitness_stat.</p>
<p><em>Note</em>: this function assumes that fitness is up to date. If you are not sure, call <a class="el" href="classhaploid__highd.html#ad5eee5bc7f757e20fb3119bb01c59b1e" title="For each clone, recalculate its traits.">update_traits()</a> and <a class="el" href="classhaploid__highd.html#a9b8ada9362e3108ee8994634231541be" title="For each clone, update fitness assuming traits are already up to date.">update_fitness()</a> first. </p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a6dd67288b3fc216d2b83d7e4861966af_icgraph.png" border="0" usemap="#classhaploid__highd_a6dd67288b3fc216d2b83d7e4861966af_icgraph" alt=""/></div>
<map name="classhaploid__highd_a6dd67288b3fc216d2b83d7e4861966af_icgraph" id="classhaploid__highd_a6dd67288b3fc216d2b83d7e4861966af_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="187,111,349,137"/><area shape="rect" id="node19" href="classhaploid__highd.html#a2dae66421820a2248db9be1771ce9e62" title="Calculate histogram of fitness from traits." alt="" coords="203,162,333,203"/><area shape="rect" id="node5" href="classhaploid__highd.html#a3351949a49ad6702beb996b58755ce4e" title="Initialize population in linkage equilibrium." alt="" coords="399,5,531,46"/><area shape="rect" id="node7" href="classhaploid__highd.html#aa0c34e02ed924603417fdb8109918f7e" title="Initialize the population with genotype counts." alt="" coords="402,70,529,111"/><area shape="rect" id="node9" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4" title="Initialize a wildtype population (00...0)" alt="" coords="402,135,529,177"/><area shape="rect" id="node17" href="classhaploid__highd.html#ae25ec86176f438db2dc0194f04ecfc42" title="Print all allele frequencies into a stream provided." alt="" coords="398,201,533,242"/><area shape="rect" id="node11" href="classhaploid__highd.html#a60617675023cab02a599a1ea7f9605c6" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="603,70,738,111"/><area shape="rect" id="node13" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="599,135,742,177"/><area shape="rect" id="node15" href="classhivpopulation.html#a299863ae882da177d548edc26d9a2cb8" title="Construct a HIV population with certain parameters." alt="" coords="581,201,760,228"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a1ee15d862374062377099130afc62c20"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::calc_individual_fitness </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structclone__t.html">clone_t</a> *&#160;</td>
          <td class="paramname"><em>tempgt</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Calculate fitness of a particular clone. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">tempgt</td><td>clone whose fitness is being calculated</td></tr>
  </table>
  </dd>
</dl>
<p>Note: this function also updates the traits information for the same clone, because the phenotype is needed to calculate fitness. If you have already calculated the traits, you can rely calc_individual_fitness_from_traits. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a1ee15d862374062377099130afc62c20_cgraph.png" border="0" usemap="#classhaploid__highd_a1ee15d862374062377099130afc62c20_cgraph" alt=""/></div>
<map name="classhaploid__highd_a1ee15d862374062377099130afc62c20_cgraph" id="classhaploid__highd_a1ee15d862374062377099130afc62c20_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="212,5,345,46"/><area shape="rect" id="node7" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="212,69,345,125"/><area shape="rect" id="node9" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="187,149,370,190"/><area shape="rect" id="node5" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="419,12,597,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a7f51d6a4f5791961d1b262e15735f820"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::calc_individual_fitness_from_traits </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structclone__t.html">clone_t</a> *&#160;</td>
          <td class="paramname"><em>tempgt</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Calculate fitness from traits of the chosen clone. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">tempgt</td><td>clone whose fitness is to be calculated </td></tr>
  </table>
  </dd>
</dl>

<p>Reimplemented in <a class="el" href="classhivpopulation.html#ae9c03725c545f943248bc32dbd5b6384">hivpopulation</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a7f51d6a4f5791961d1b262e15735f820_icgraph.png" border="0" usemap="#classhaploid__highd_a7f51d6a4f5791961d1b262e15735f820_icgraph" alt=""/></div>
<map name="classhaploid__highd_a7f51d6a4f5791961d1b262e15735f820_icgraph" id="classhaploid__highd_a7f51d6a4f5791961d1b262e15735f820_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a943354008a9931e4223f952f4790efd6" title="Flip a spin (locus) in individual." alt="" coords="209,57,335,98"/><area shape="rect" id="node11" href="classhaploid__highd.html#afb3d9e695f16b8412902757f18cfb63e" title="Recombine two genotypes parent1 and parent2 to produce two new genotypes." alt="" coords="187,136,357,163"/><area shape="rect" id="node16" href="classhaploid__highd.html#a9b8ada9362e3108ee8994634231541be" title="For each clone, update fitness assuming traits are already up to date." alt="" coords="198,285,346,326"/><area shape="rect" id="node40" href="classhaploid__highd.html#a1ee15d862374062377099130afc62c20" title="Calculate fitness of a particular clone." alt="" coords="205,350,339,391"/><area shape="rect" id="node42" href="classhaploid__highd.html#a44235ff9f17f4e81f290611d7829818a" title="Add the genotype specified by a bitset to the current population in in n copies." alt="" coords="421,187,552,229"/><area shape="rect" id="node5" href="classhaploid__highd.html#af173582ee41aecd03d8f97aa234cf879" title="Mutate random clones at random loci (all loci)" alt="" coords="412,5,561,32"/><area shape="rect" id="node9" href="classhaploid__highd.html#a710a61fe8975aaaf4d148b2da20985ae" title="Flip a spin at a specific locus in random individual." alt="" coords="424,57,549,98"/><area shape="rect" id="node7" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="617,49,759,76"/><area shape="rect" id="node13" href="classhaploid__highd.html#ab08783cb783ca66ff2c87227dbd74314" title="Pair and mate sexual gametes." alt="" coords="421,122,552,163"/><area shape="rect" id="node18" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="405,449,568,476"/><area shape="rect" id="node34" href="classhivpopulation.html#ad1caa476f2d93020d52d0ce46f81a529" title="hivpopulation::read\l_replication_coefficients" alt="" coords="407,318,567,359"/><area shape="rect" id="node36" href="classhivpopulation.html#a6afaadd90cf25378476a13a2bef80875" title="hivpopulation::read\l_resistance_coefficients" alt="" coords="406,383,567,425"/><area shape="rect" id="node38" href="classhivpopulation.html#a0f5bd29413e44f8a075afe164c05853e" title="hivpopulation::set\l_treatment" alt="" coords="425,253,548,294"/><area shape="rect" id="node20" href="classhaploid__highd.html#a3351949a49ad6702beb996b58755ce4e" title="Initialize population in linkage equilibrium." alt="" coords="622,311,754,353"/><area shape="rect" id="node22" href="classhaploid__highd.html#aa0c34e02ed924603417fdb8109918f7e" title="Initialize the population with genotype counts." alt="" coords="625,246,751,287"/><area shape="rect" id="node24" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4" title="Initialize a wildtype population (00...0)" alt="" coords="625,377,751,418"/><area shape="rect" id="node32" href="classhaploid__highd.html#ae25ec86176f438db2dc0194f04ecfc42" title="Print all allele frequencies into a stream provided." alt="" coords="621,442,755,483"/><area shape="rect" id="node26" href="classhaploid__highd.html#a60617675023cab02a599a1ea7f9605c6" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="830,289,965,330"/><area shape="rect" id="node28" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="826,166,969,207"/><area shape="rect" id="node30" href="classhivpopulation.html#a299863ae882da177d548edc26d9a2cb8" title="Construct a HIV population with certain parameters." alt="" coords="808,384,987,411"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="acf40ae60f1441f5a8bb67a5a84444c01"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::calc_individual_traits </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structclone__t.html">clone_t</a> *&#160;</td>
          <td class="paramname"><em>tempgt</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Calculate traits of the chosen clone. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">tempgt</td><td>clone whose traits are to be calculated </td></tr>
  </table>
  </dd>
</dl>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_acf40ae60f1441f5a8bb67a5a84444c01_cgraph.png" border="0" usemap="#classhaploid__highd_acf40ae60f1441f5a8bb67a5a84444c01_cgraph" alt=""/></div>
<map name="classhaploid__highd_acf40ae60f1441f5a8bb67a5a84444c01_cgraph" id="classhaploid__highd_acf40ae60f1441f5a8bb67a5a84444c01_cgraph">
<area shape="rect" id="node3" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="187,12,365,39"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_acf40ae60f1441f5a8bb67a5a84444c01_icgraph.png" border="0" usemap="#classhaploid__highd_acf40ae60f1441f5a8bb67a5a84444c01_icgraph" alt=""/></div>
<map name="classhaploid__highd_acf40ae60f1441f5a8bb67a5a84444c01_icgraph" id="classhaploid__highd_acf40ae60f1441f5a8bb67a5a84444c01_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#afb3d9e695f16b8412902757f18cfb63e" title="Recombine two genotypes parent1 and parent2 to produce two new genotypes." alt="" coords="187,91,357,117"/><area shape="rect" id="node9" href="classhaploid__highd.html#ad5eee5bc7f757e20fb3119bb01c59b1e" title="For each clone, recalculate its traits." alt="" coords="198,223,346,265"/><area shape="rect" id="node33" href="classhaploid__highd.html#a1ee15d862374062377099130afc62c20" title="Calculate fitness of a particular clone." alt="" coords="205,289,339,330"/><area shape="rect" id="node35" href="classhaploid__highd.html#a44235ff9f17f4e81f290611d7829818a" title="Add the genotype specified by a bitset to the current population in in n copies." alt="" coords="421,129,552,170"/><area shape="rect" id="node5" href="classhaploid__highd.html#ab08783cb783ca66ff2c87227dbd74314" title="Pair and mate sexual gametes." alt="" coords="421,11,552,53"/><area shape="rect" id="node7" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="617,5,759,32"/><area shape="rect" id="node11" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="405,260,568,287"/><area shape="rect" id="node27" href="classhivpopulation.html#ad1caa476f2d93020d52d0ce46f81a529" title="hivpopulation::read\l_replication_coefficients" alt="" coords="407,377,567,418"/><area shape="rect" id="node29" href="classhivpopulation.html#a6afaadd90cf25378476a13a2bef80875" title="hivpopulation::read\l_resistance_coefficients" alt="" coords="406,194,567,235"/><area shape="rect" id="node31" href="classhivpopulation.html#a0f5bd29413e44f8a075afe164c05853e" title="hivpopulation::set\l_treatment" alt="" coords="425,311,548,353"/><area shape="rect" id="node13" href="classhaploid__highd.html#a3351949a49ad6702beb996b58755ce4e" title="Initialize population in linkage equilibrium." alt="" coords="622,223,754,265"/><area shape="rect" id="node15" href="classhaploid__highd.html#aa0c34e02ed924603417fdb8109918f7e" title="Initialize the population with genotype counts." alt="" coords="625,158,751,199"/><area shape="rect" id="node17" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4" title="Initialize a wildtype population (00...0)" alt="" coords="625,289,751,330"/><area shape="rect" id="node25" href="classhaploid__highd.html#ae25ec86176f438db2dc0194f04ecfc42" title="Print all allele frequencies into a stream provided." alt="" coords="621,354,755,395"/><area shape="rect" id="node19" href="classhaploid__highd.html#a60617675023cab02a599a1ea7f9605c6" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="830,201,965,242"/><area shape="rect" id="node21" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="826,78,969,119"/><area shape="rect" id="node23" href="classhivpopulation.html#a299863ae882da177d548edc26d9a2cb8" title="Construct a HIV population with certain parameters." alt="" coords="808,296,987,323"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="af40993ccd48558777ab65ab88bc590b3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::calc_stat </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculate trait and fitness statistics and allele frequences. </p>
<p>Four things are done in a row:</p>
<ol type="1">
<li>traits are updated</li>
<li>fitness is updated based on the new traits</li>
<li>statistics of traits and fitness are calculated</li>
<li>allele freqs are calculated</li>
</ol>
<p>Note: This function is quite expensive. Please use its subblocks whenever possible, and rely on this only when you want to make sure, at the expense of performance, that everything is up to date. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_af40993ccd48558777ab65ab88bc590b3_cgraph.png" border="0" usemap="#classhaploid__highd_af40993ccd48558777ab65ab88bc590b3_cgraph" alt=""/></div>
<map name="classhaploid__highd_af40993ccd48558777ab65ab88bc590b3_cgraph" id="classhaploid__highd_af40993ccd48558777ab65ab88bc590b3_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ad5eee5bc7f757e20fb3119bb01c59b1e" title="For each clone, recalculate its traits." alt="" coords="217,9,365,50"/><area shape="rect" id="node9" href="classhaploid__highd.html#a9b8ada9362e3108ee8994634231541be" title="For each clone, update fitness assuming traits are already up to date." alt="" coords="217,77,365,118"/><area shape="rect" id="node15" href="classhaploid__highd.html#afb3de45e613ffe03eadcc10f81c13448" title="Calculate and store trait population statistics and covariances." alt="" coords="224,142,357,183"/><area shape="rect" id="node17" href="classhaploid__highd.html#a6dd67288b3fc216d2b83d7e4861966af" title="Calculate and store fitness population statistics." alt="" coords="224,207,357,249"/><area shape="rect" id="node19" href="classhaploid__highd.html#a93cb185958cc78ce66c909d8df532a57" title="calculate and store allele frequencies" alt="" coords="224,273,357,314"/><area shape="rect" id="node5" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="439,5,572,46"/><area shape="rect" id="node7" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="646,12,823,39"/><area shape="rect" id="node11" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="439,69,572,125"/><area shape="rect" id="node13" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="414,149,597,190"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_af40993ccd48558777ab65ab88bc590b3_icgraph.png" border="0" usemap="#classhaploid__highd_af40993ccd48558777ab65ab88bc590b3_icgraph" alt=""/></div>
<map name="classhaploid__highd_af40993ccd48558777ab65ab88bc590b3_icgraph" id="classhaploid__highd_af40993ccd48558777ab65ab88bc590b3_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a3351949a49ad6702beb996b58755ce4e" title="Initialize population in linkage equilibrium." alt="" coords="218,5,350,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#aa0c34e02ed924603417fdb8109918f7e" title="Initialize the population with genotype counts." alt="" coords="221,70,347,111"/><area shape="rect" id="node7" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4" title="Initialize a wildtype population (00...0)" alt="" coords="221,135,347,177"/><area shape="rect" id="node15" href="classhaploid__highd.html#ae25ec86176f438db2dc0194f04ecfc42" title="Print all allele frequencies into a stream provided." alt="" coords="217,201,351,242"/><area shape="rect" id="node9" href="classhaploid__highd.html#a60617675023cab02a599a1ea7f9605c6" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="422,70,557,111"/><area shape="rect" id="node11" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="418,135,561,177"/><area shape="rect" id="node13" href="classhivpopulation.html#a299863ae882da177d548edc26d9a2cb8" title="Construct a HIV population with certain parameters." alt="" coords="400,201,579,228"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="afb3de45e613ffe03eadcc10f81c13448"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::calc_trait_stat </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Calculate and store trait population statistics and covariances. </p>
<p>The traits statistics are stored in traits_stat.</p>
<p><em>Note</em>: this function assumes that traits are up to date. If you are not sure, call <a class="el" href="classhaploid__highd.html#ad5eee5bc7f757e20fb3119bb01c59b1e" title="For each clone, recalculate its traits.">update_traits()</a> first. </p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_afb3de45e613ffe03eadcc10f81c13448_icgraph.png" border="0" usemap="#classhaploid__highd_afb3de45e613ffe03eadcc10f81c13448_icgraph" alt=""/></div>
<map name="classhaploid__highd_afb3de45e613ffe03eadcc10f81c13448_icgraph" id="classhaploid__highd_afb3de45e613ffe03eadcc10f81c13448_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="187,111,349,137"/><area shape="rect" id="node5" href="classhaploid__highd.html#a3351949a49ad6702beb996b58755ce4e" title="Initialize population in linkage equilibrium." alt="" coords="399,5,531,46"/><area shape="rect" id="node7" href="classhaploid__highd.html#aa0c34e02ed924603417fdb8109918f7e" title="Initialize the population with genotype counts." alt="" coords="402,70,529,111"/><area shape="rect" id="node9" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4" title="Initialize a wildtype population (00...0)" alt="" coords="402,135,529,177"/><area shape="rect" id="node17" href="classhaploid__highd.html#ae25ec86176f438db2dc0194f04ecfc42" title="Print all allele frequencies into a stream provided." alt="" coords="398,201,533,242"/><area shape="rect" id="node11" href="classhaploid__highd.html#a60617675023cab02a599a1ea7f9605c6" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="603,70,738,111"/><area shape="rect" id="node13" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="599,135,742,177"/><area shape="rect" id="node15" href="classhivpopulation.html#a299863ae882da177d548edc26d9a2cb8" title="Construct a HIV population with certain parameters." alt="" coords="581,201,760,228"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a0d929c061bbf1c2fd4d82f6bb2f2f998"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::check_individual_maximal_fitness </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structclone__t.html">clone_t</a> *&#160;</td>
          <td class="paramname"><em>tempgt</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a0d929c061bbf1c2fd4d82f6bb2f2f998_icgraph.png" border="0" usemap="#classhaploid__highd_a0d929c061bbf1c2fd4d82f6bb2f2f998_icgraph" alt=""/></div>
<map name="classhaploid__highd_a0d929c061bbf1c2fd4d82f6bb2f2f998_icgraph" id="classhaploid__highd_a0d929c061bbf1c2fd4d82f6bb2f2f998_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a943354008a9931e4223f952f4790efd6" title="Flip a spin (locus) in individual." alt="" coords="260,57,385,98"/><area shape="rect" id="node11" href="classhaploid__highd.html#afb3d9e695f16b8412902757f18cfb63e" title="Recombine two genotypes parent1 and parent2 to produce two new genotypes." alt="" coords="238,136,407,163"/><area shape="rect" id="node16" href="classhaploid__highd.html#a9b8ada9362e3108ee8994634231541be" title="For each clone, update fitness assuming traits are already up to date." alt="" coords="249,285,397,326"/><area shape="rect" id="node40" href="classhaploid__highd.html#a1ee15d862374062377099130afc62c20" title="Calculate fitness of a particular clone." alt="" coords="256,350,389,391"/><area shape="rect" id="node42" href="classhaploid__highd.html#a44235ff9f17f4e81f290611d7829818a" title="Add the genotype specified by a bitset to the current population in in n copies." alt="" coords="472,187,603,229"/><area shape="rect" id="node5" href="classhaploid__highd.html#af173582ee41aecd03d8f97aa234cf879" title="Mutate random clones at random loci (all loci)" alt="" coords="463,5,612,32"/><area shape="rect" id="node9" href="classhaploid__highd.html#a710a61fe8975aaaf4d148b2da20985ae" title="Flip a spin at a specific locus in random individual." alt="" coords="475,57,600,98"/><area shape="rect" id="node7" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="667,49,810,76"/><area shape="rect" id="node13" href="classhaploid__highd.html#ab08783cb783ca66ff2c87227dbd74314" title="Pair and mate sexual gametes." alt="" coords="472,122,603,163"/><area shape="rect" id="node18" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="456,449,619,476"/><area shape="rect" id="node34" href="classhivpopulation.html#ad1caa476f2d93020d52d0ce46f81a529" title="hivpopulation::read\l_replication_coefficients" alt="" coords="457,318,617,359"/><area shape="rect" id="node36" href="classhivpopulation.html#a6afaadd90cf25378476a13a2bef80875" title="hivpopulation::read\l_resistance_coefficients" alt="" coords="457,383,618,425"/><area shape="rect" id="node38" href="classhivpopulation.html#a0f5bd29413e44f8a075afe164c05853e" title="hivpopulation::set\l_treatment" alt="" coords="476,253,599,294"/><area shape="rect" id="node20" href="classhaploid__highd.html#a3351949a49ad6702beb996b58755ce4e" title="Initialize population in linkage equilibrium." alt="" coords="673,311,805,353"/><area shape="rect" id="node22" href="classhaploid__highd.html#aa0c34e02ed924603417fdb8109918f7e" title="Initialize the population with genotype counts." alt="" coords="675,246,802,287"/><area shape="rect" id="node24" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4" title="Initialize a wildtype population (00...0)" alt="" coords="675,377,802,418"/><area shape="rect" id="node32" href="classhaploid__highd.html#ae25ec86176f438db2dc0194f04ecfc42" title="Print all allele frequencies into a stream provided." alt="" coords="671,442,806,483"/><area shape="rect" id="node26" href="classhaploid__highd.html#a60617675023cab02a599a1ea7f9605c6" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="881,289,1015,330"/><area shape="rect" id="node28" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="877,166,1019,207"/><area shape="rect" id="node30" href="classhivpopulation.html#a299863ae882da177d548edc26d9a2cb8" title="Construct a HIV population with certain parameters." alt="" coords="859,384,1037,411"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a298bdd5103be21db9f72fb72685cd414"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::clear_fitness </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a0a5523161a38427a7bdcc18136751817"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::clear_trait </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>t</em> = <code>0</code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="acc3e232638ff352b52a4c8db9b729a23"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::clear_traits </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a33a62c8a5c57d36e071099906131eb31"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">boost::dynamic_bitset haploid_highd::crossover_pattern </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Choose a number of crossover points and produce a crossover pattern. </p>
<dl class="section return"><dt>Returns</dt><dd>crossover pattern</dd></dl>
<p>A typical crossover pattern would be 0000111100011111101101. </p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a33a62c8a5c57d36e071099906131eb31_icgraph.png" border="0" usemap="#classhaploid__highd_a33a62c8a5c57d36e071099906131eb31_icgraph" alt=""/></div>
<map name="classhaploid__highd_a33a62c8a5c57d36e071099906131eb31_icgraph" id="classhaploid__highd_a33a62c8a5c57d36e071099906131eb31_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#afb3d9e695f16b8412902757f18cfb63e" title="Recombine two genotypes parent1 and parent2 to produce two new genotypes." alt="" coords="219,12,389,39"/><area shape="rect" id="node5" href="classhaploid__highd.html#ab08783cb783ca66ff2c87227dbd74314" title="Pair and mate sexual gametes." alt="" coords="437,5,568,46"/><area shape="rect" id="node7" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="617,12,759,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a686aa6ae9c4d31f96dd6bb3d83d11188"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::distance_Hamming </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>clone1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>clone2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; unsigned int * &gt; *&#160;</td>
          <td class="paramname"><em>chunks</em> = <code>NULL</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>every</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a686aa6ae9c4d31f96dd6bb3d83d11188_cgraph.png" border="0" usemap="#classhaploid__highd_a686aa6ae9c4d31f96dd6bb3d83d11188_cgraph" alt=""/></div>
<map name="classhaploid__highd_a686aa6ae9c4d31f96dd6bb3d83d11188_cgraph" id="classhaploid__highd_a686aa6ae9c4d31f96dd6bb3d83d11188_cgraph">
</map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a686aa6ae9c4d31f96dd6bb3d83d11188_icgraph.png" border="0" usemap="#classhaploid__highd_a686aa6ae9c4d31f96dd6bb3d83d11188_icgraph" alt=""/></div>
<map name="classhaploid__highd_a686aa6ae9c4d31f96dd6bb3d83d11188_icgraph" id="classhaploid__highd_a686aa6ae9c4d31f96dd6bb3d83d11188_icgraph">
<area shape="rect" id="node4" href="classhaploid__highd.html#abf5d073bfa998684b94c0b3248d7bdad" title="Calculate diversity in the current population (Hamming distance between pairs of sequences)" alt="" coords="223,5,359,46"/><area shape="rect" id="node6" href="classhaploid__highd.html#a234a418eff9319a2dfb6371b73220b56" title="Get histogram of divergence from the [00...0] bitset." alt="" coords="213,70,368,111"/><area shape="rect" id="node8" href="classhaploid__highd.html#a25a490ade017cff4cccb88ced04261ce" title="Get histogram of diversity in the population (mutual Hamming distance)" alt="" coords="221,135,361,177"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aab7c4bf82bda5140c246b7d55cd2906d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::distance_Hamming </td>
          <td>(</td>
          <td class="paramtype">boost::dynamic_bitset&lt;&gt;&#160;</td>
          <td class="paramname"><em>gt1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">boost::dynamic_bitset&lt;&gt;&#160;</td>
          <td class="paramname"><em>gt2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; unsigned int * &gt; *&#160;</td>
          <td class="paramname"><em>chunks</em> = <code>NULL</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>every</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculate Hamming distance between two sequences. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">gt1</td><td>first sequence </td></tr>
    <tr><td class="paramname">gt2</td><td>second sequence </td></tr>
    <tr><td class="paramname">chunks</td><td>(pointer to) vector of ranges (C pairs), e.g. ((0,10), (25, 28), (68, 70)) </td></tr>
    <tr><td class="paramname">every</td><td>check only every X sites, starting from the first of each chunk</td></tr>
  </table>
  </dd>
</dl>
<p><em>Note</em>: you cannot use <code>every</code> without <code>chunks</code>.</p>
<p><em>Note</em>: every chunk is a C array of length 2, i.e. a pointer to lower boundary of the interval. Thus we have the following situation:</p>
<ul>
<li><code>chunks</code> is of type vector <code>&lt;unsigned int *&gt; *</code></li>
<li><code>*chunks</code> is of type vector <code>&lt;unsigned int *&gt;</code></li>
<li><code>(*chunks)[i]</code> is of type <code>*unsigned int</code></li>
<li><code>(*chunks)[i][0]</code> is of type <code>unsigned int</code>, and indicates the initial site for this chunk</li>
<li><code>(*chunks)[i][1]</code> is of type <code>unsigned int</code>, and indicates the (final site +1) for this chunk</li>
</ul>
<p>As a consequence, for each chunk, chunk[1] - chunk[0] is the length of the chunk.</p>
<dl class="section return"><dt>Returns</dt><dd>Hamming distance, not normalized</dd></dl>
<p>When you prepare the vector of chunks for this function, please use C++ methods (<code>new</code>) that take care of memory management, or be very careful about memory leaks.</p>
<p><em>Note</em>: this function is overloaded with simpler arguments (e.g. if you want to use clone indices). </p>

</div>
</div>
<a class="anchor" id="af7fbfc38495246eb5ad91201c9a8f58a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::evolve </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>gen</em> = <code>1</code></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Evolve for some generations under the specified conditions. </p>
<p>The order of steps performed is</p>
<ol type="1">
<li>mutation</li>
<li>selection</li>
<li>recombination but should not be important except for extremely high selection coefficients and/or very short times, for which the discrete nature of the Fisher-Wright model becomes relevant.</li>
</ol>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">gen</td><td>number of generations.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>zero if successful, error code in the faulty step otherwise</dd></dl>
<p>Note: if an error in encountered, evolution is stopped after the function that created the problem. Typical errors include extinction or the opposite, offspring explosion. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_af7fbfc38495246eb5ad91201c9a8f58a_cgraph.png" border="0" usemap="#classhaploid__highd_af7fbfc38495246eb5ad91201c9a8f58a_cgraph" alt=""/></div>
<map name="classhaploid__highd_af7fbfc38495246eb5ad91201c9a8f58a_cgraph" id="classhaploid__highd_af7fbfc38495246eb5ad91201c9a8f58a_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af173582ee41aecd03d8f97aa234cf879" title="Mutate random clones at random loci (all loci)" alt="" coords="197,112,347,139"/><area shape="rect" id="node22" href="classhaploid__highd.html#aec2c31155f262a9c1cf21262eea2d03a" title="Generate offspring according to fitness (selection) and segregate some for sexual mating..." alt="" coords="200,345,344,386"/><area shape="rect" id="node28" href="classhaploid__highd.html#ab08783cb783ca66ff2c87227dbd74314" title="Pair and mate sexual gametes." alt="" coords="207,279,337,321"/><area shape="rect" id="node42" href="classhaploid__highd.html#a6fa4ffe295eef314642dec0208f81b7d" title="Make the the temporary population the current one." alt="" coords="203,410,341,451"/><area shape="rect" id="node5" href="classhaploid__highd.html#ab3e442c79f83e603bc30892e2a36e329" title="Produce and store a random sample of the population for stochastic processes." alt="" coords="629,5,784,46"/><area shape="rect" id="node7" href="classhaploid__highd.html#a943354008a9931e4223f952f4790efd6" title="Flip a spin (locus) in individual." alt="" coords="417,141,543,182"/><area shape="rect" id="node19" href="classhaploid__highd.html#ab64531c02f245f39d9446efb5fcabe01" title="Get a random clone from the population." alt="" coords="404,69,556,110"/><area shape="rect" id="node9" href="classhaploid__highd.html#ac4d5cd6a0e86ae30257f41c1e5b889c5" title="Calculate trait difference between two clones." alt="" coords="643,70,770,111"/><area shape="rect" id="node15" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="640,135,773,191"/><area shape="rect" id="node17" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="615,214,798,255"/><area shape="rect" id="node11" href="classhypercube__highd.html#a8352b2198e0b8f6d352face30dff6989" title="Calculate difference between two hypercube points efficiently." alt="" coords="864,70,1011,111"/><area shape="rect" id="node13" href="classhypercube__highd.html#a44781cd1dcf006cd33263ca5fbdf245a" title=": get the trait coefficient of a locus." alt="" coords="1075,70,1221,111"/><area shape="rect" id="node24" href="classhaploid__highd.html#ae64df329db764e98e77398a775a43600" title="Get the log of the exp&#45;average fitness plus relaxation term." alt="" coords="397,410,563,451"/><area shape="rect" id="node26" href="classhaploid__highd.html#a0aaa82167545768f814c7e28868eee77" title="Get the population exp&#45;average of fitness, used for keeping the population size fixed." alt="" coords="635,475,779,517"/><area shape="rect" id="node30" href="classhaploid__highd.html#afb3d9e695f16b8412902757f18cfb63e" title="Recombine two genotypes parent1 and parent2 to produce two new genotypes." alt="" coords="395,287,565,313"/><area shape="rect" id="node32" href="classhaploid__highd.html#ac79972f0e845b52b3191a27ddae309a8" title="Produce a random reassortement pattern." alt="" coords="614,279,799,321"/><area shape="rect" id="node34" href="classhaploid__highd.html#a33a62c8a5c57d36e071099906131eb31" title="Choose a number of crossover points and produce a crossover pattern." alt="" coords="625,345,789,386"/><area shape="rect" id="node36" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="640,410,773,451"/><area shape="rect" id="node38" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="849,417,1026,444"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a943354008a9931e4223f952f4790efd6"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::flip_single_locus </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>clonenum</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>locus</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Flip a spin (locus) in individual. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">clonenum</td><td>the individual whose locus is being flipped </td></tr>
    <tr><td class="paramname">locus</td><td>position of the locus to flip</td></tr>
  </table>
  </dd>
</dl>
<p>This function creates a new clone and adds it to the population, and assigns it a fitness.</p>
<p>Note: This might produce duplicate clones since the mutant clone produced might already exist. Duplicates can be merged by the member <a class="el" href="classhaploid__highd.html#a981292b7434288cbee87093576b61371" title="Remove duplicate clones.">unique_clones()</a> </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a943354008a9931e4223f952f4790efd6_cgraph.png" border="0" usemap="#classhaploid__highd_a943354008a9931e4223f952f4790efd6_cgraph" alt=""/></div>
<map name="classhaploid__highd_a943354008a9931e4223f952f4790efd6_cgraph" id="classhaploid__highd_a943354008a9931e4223f952f4790efd6_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ac4d5cd6a0e86ae30257f41c1e5b889c5" title="Calculate trait difference between two clones." alt="" coords="207,5,334,46"/><area shape="rect" id="node9" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="204,69,337,125"/><area shape="rect" id="node11" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="179,149,362,190"/><area shape="rect" id="node5" href="classhypercube__highd.html#a8352b2198e0b8f6d352face30dff6989" title="Calculate difference between two hypercube points efficiently." alt="" coords="411,5,557,46"/><area shape="rect" id="node7" href="classhypercube__highd.html#a44781cd1dcf006cd33263ca5fbdf245a" title=": get the trait coefficient of a locus." alt="" coords="605,5,752,46"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a943354008a9931e4223f952f4790efd6_icgraph.png" border="0" usemap="#classhaploid__highd_a943354008a9931e4223f952f4790efd6_icgraph" alt=""/></div>
<map name="classhaploid__highd_a943354008a9931e4223f952f4790efd6_icgraph" id="classhaploid__highd_a943354008a9931e4223f952f4790efd6_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af173582ee41aecd03d8f97aa234cf879" title="Mutate random clones at random loci (all loci)" alt="" coords="179,5,328,32"/><area shape="rect" id="node7" href="classhaploid__highd.html#a710a61fe8975aaaf4d148b2da20985ae" title="Flip a spin at a specific locus in random individual." alt="" coords="191,57,316,98"/><area shape="rect" id="node5" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="377,5,519,32"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a710a61fe8975aaaf4d148b2da20985ae"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::flip_single_locus </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>locus</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Flip a spin at a specific locus in random individual. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">locus</td><td>position of the locus to flip</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the (random) flipped individual clone</dd></dl>
<p>Note: this function calls <a class="el" href="classhaploid__highd.html#a943354008a9931e4223f952f4790efd6" title="Flip a spin (locus) in individual.">flip_single_locus(unsigned int clonenum, int locus)</a>. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a710a61fe8975aaaf4d148b2da20985ae_cgraph.png" border="0" usemap="#classhaploid__highd_a710a61fe8975aaaf4d148b2da20985ae_cgraph" alt=""/></div>
<map name="classhaploid__highd_a710a61fe8975aaaf4d148b2da20985ae_cgraph" id="classhaploid__highd_a710a61fe8975aaaf4d148b2da20985ae_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ab64531c02f245f39d9446efb5fcabe01" title="Get a random clone from the population." alt="" coords="179,23,331,65"/><area shape="rect" id="node7" href="classhaploid__highd.html#a943354008a9931e4223f952f4790efd6" title="Flip a spin (locus) in individual." alt="" coords="192,125,317,166"/><area shape="rect" id="node5" href="classhaploid__highd.html#ab3e442c79f83e603bc30892e2a36e329" title="Produce and store a random sample of the population for stochastic processes." alt="" coords="393,5,548,46"/><area shape="rect" id="node9" href="classhaploid__highd.html#ac4d5cd6a0e86ae30257f41c1e5b889c5" title="Calculate trait difference between two clones." alt="" coords="407,70,534,111"/><area shape="rect" id="node15" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="404,135,537,191"/><area shape="rect" id="node17" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="379,214,562,255"/><area shape="rect" id="node11" href="classhypercube__highd.html#a8352b2198e0b8f6d352face30dff6989" title="Calculate difference between two hypercube points efficiently." alt="" coords="611,70,757,111"/><area shape="rect" id="node13" href="classhypercube__highd.html#a44781cd1dcf006cd33263ca5fbdf245a" title=": get the trait coefficient of a locus." alt="" coords="805,70,952,111"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a0c1eb3defc6dbf00e414482ae39355e2"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_allele_frequency </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>l</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ad94f599d9a5c7e3a13afe96a0a240d73"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_chi </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>l</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ae87baa777bd8550c0d80ca44397a35b7"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_chi2 </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>locus1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>locus2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a8351db0129fbfb98170f0b60565c9dc3"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::get_clone_size </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a234a418eff9319a2dfb6371b73220b56"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::get_divergence_histogram </td>
          <td>(</td>
          <td class="paramtype">gsl_histogram **&#160;</td>
          <td class="paramname"><em>hist</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>bins</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; unsigned int * &gt; *&#160;</td>
          <td class="paramname"><em>chunks</em> = <code>NULL</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>every</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>n_sample</em> = <code>1000</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get histogram of divergence from the [00...0] bitset. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">hist</td><td>pointer to the gsl_histogram to fill </td></tr>
    <tr><td class="paramname">bins</td><td>number of bins in the histogram</td></tr>
  </table>
  </dd>
</dl>
<p><em>Note</em>: an antialiasing algorithm is used to enforce integer bin widths. <em>Note</em>: because of the antialiasing requirement, the last bin might be underrepresented. <em>Note</em>: the number of used bins in the output might be smaller than requested, depending on sample size and antialiasing.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">chunks</td><td>(pointer to) vector of ranges (C pairs), e.g. ((0,10), (25, 28), (68, 70)) </td></tr>
    <tr><td class="paramname">every</td><td>check only every X sites, starting from the first of each chunk</td></tr>
  </table>
  </dd>
</dl>
<p><em>Note</em>: you cannot use <code>every</code> without <code>chunks</code>.</p>
<p><em>Note</em>: every chunk is a C array of length 2, i.e. a pointer to lower boundary of the interval. Thus we have the following situation:</p>
<ul>
<li><code>chunks</code> is of type vector <code>&lt;unsigned int *&gt; *</code></li>
<li><code>*chunks</code> is of type vector <code>&lt;unsigned int *&gt;</code></li>
<li><code>(*chunks)[i]</code> is of type <code>*unsigned int</code></li>
<li><code>(*chunks)[i][0]</code> is of type <code>unsigned int</code>, and indicates the initial site for this chunk</li>
<li><code>(*chunks)[i][1]</code> is of type <code>unsigned int</code>, and indicates the (final site +1) for this chunk</li>
</ul>
<p>As a consequence, for each chunk, chunk[1] - chunk[0] is the length of the chunk.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n_sample</td><td>size of the random sample to use (the whole population is often too large)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>zero if successful, error codes otherwise </dd></dl>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a234a418eff9319a2dfb6371b73220b56_cgraph.png" border="0" usemap="#classhaploid__highd_a234a418eff9319a2dfb6371b73220b56_cgraph" alt=""/></div>
<map name="classhaploid__highd_a234a418eff9319a2dfb6371b73220b56_cgraph" id="classhaploid__highd_a234a418eff9319a2dfb6371b73220b56_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ab3e442c79f83e603bc30892e2a36e329" title="Produce and store a random sample of the population for stochastic processes." alt="" coords="616,5,771,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#aefe07878a3133687465c41804043779d" title="Sample random individuals from the population." alt="" coords="212,34,364,75"/><area shape="rect" id="node10" href="classhaploid__highd.html#a686aa6ae9c4d31f96dd6bb3d83d11188" title="haploid_highd::distance\l_Hamming" alt="" coords="209,123,367,165"/><area shape="rect" id="node7" href="classhaploid__highd.html#ab64531c02f245f39d9446efb5fcabe01" title="Get a random clone from the population." alt="" coords="416,34,568,75"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="af77958954fe6001b2640b70bcc3681a4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structstat__t.html">stat_t</a> haploid_highd::get_divergence_statistics </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>n_sample</em> = <code>1000</code></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculate mean and variance of the divergence from the [00...0] bitset. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n_sample</td><td>size of the statistical sample to use (the whole pop is often too large)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>mean and variance of the divergence in a <a class="el" href="structstat__t.html" title="Structure for short summary statistics.">stat_t</a> </dd></dl>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_af77958954fe6001b2640b70bcc3681a4_cgraph.png" border="0" usemap="#classhaploid__highd_af77958954fe6001b2640b70bcc3681a4_cgraph" alt=""/></div>
<map name="classhaploid__highd_af77958954fe6001b2640b70bcc3681a4_cgraph" id="classhaploid__highd_af77958954fe6001b2640b70bcc3681a4_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ab3e442c79f83e603bc30892e2a36e329" title="Produce and store a random sample of the population for stochastic processes." alt="" coords="605,6,760,47"/><area shape="rect" id="node5" href="classhaploid__highd.html#aefe07878a3133687465c41804043779d" title="Sample random individuals from the population." alt="" coords="205,35,357,77"/><area shape="rect" id="node7" href="classhaploid__highd.html#ab64531c02f245f39d9446efb5fcabe01" title="Get a random clone from the population." alt="" coords="405,35,557,77"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a25a490ade017cff4cccb88ced04261ce"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::get_diversity_histogram </td>
          <td>(</td>
          <td class="paramtype">gsl_histogram **&#160;</td>
          <td class="paramname"><em>hist</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>bins</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; unsigned int * &gt; *&#160;</td>
          <td class="paramname"><em>chunks</em> = <code>NULL</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>every</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>n_sample</em> = <code>1000</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get histogram of diversity in the population (mutual Hamming distance) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">hist</td><td>pointer to the gsl_histogram to fill </td></tr>
    <tr><td class="paramname">bins</td><td>number of bins in the histogram</td></tr>
  </table>
  </dd>
</dl>
<p><em>Note</em>: an antialiasing algorithm is used to enforce integer bin widths. <em>Note</em>: because of the antialiasing requirement, the last bin might be underrepresented. <em>Note</em>: the number of used bins in the output might be smaller than requested, depending on sample size and antialiasing.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">chunks</td><td>(pointer to) vector of ranges (C pairs), e.g. ((0,10), (25, 28), (68, 70)) </td></tr>
    <tr><td class="paramname">every</td><td>check only every X sites, starting from the first of each chunk</td></tr>
  </table>
  </dd>
</dl>
<p><em>Note</em>: you cannot use <code>every</code> without <code>chunks</code>.</p>
<p><em>Note</em>: every chunk is a C array of length 2, i.e. a pointer to lower boundary of the interval. Thus we have the following situation:</p>
<ul>
<li><code>chunks</code> is of type vector <code>&lt;unsigned int *&gt; *</code></li>
<li><code>*chunks</code> is of type vector <code>&lt;unsigned int *&gt;</code></li>
<li><code>(*chunks)[i]</code> is of type <code>*unsigned int</code></li>
<li><code>(*chunks)[i][0]</code> is of type <code>unsigned int</code>, and indicates the initial site for this chunk</li>
<li><code>(*chunks)[i][1]</code> is of type <code>unsigned int</code>, and indicates the (final site +1) for this chunk</li>
</ul>
<p>As a consequence, for each chunk, chunk[1] - chunk[0] is the length of the chunk.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n_sample</td><td>size of the random sample to use (the whole population is often too large)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>zero if successful, error codes otherwise </dd></dl>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a25a490ade017cff4cccb88ced04261ce_cgraph.png" border="0" usemap="#classhaploid__highd_a25a490ade017cff4cccb88ced04261ce_cgraph" alt=""/></div>
<map name="classhaploid__highd_a25a490ade017cff4cccb88ced04261ce_cgraph" id="classhaploid__highd_a25a490ade017cff4cccb88ced04261ce_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ab3e442c79f83e603bc30892e2a36e329" title="Produce and store a random sample of the population for stochastic processes." alt="" coords="603,5,757,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#aefe07878a3133687465c41804043779d" title="Sample random individuals from the population." alt="" coords="199,34,351,75"/><area shape="rect" id="node10" href="classhaploid__highd.html#a686aa6ae9c4d31f96dd6bb3d83d11188" title="haploid_highd::distance\l_Hamming" alt="" coords="195,123,354,165"/><area shape="rect" id="node7" href="classhaploid__highd.html#ab64531c02f245f39d9446efb5fcabe01" title="Get a random clone from the population." alt="" coords="403,34,555,75"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="abf5d073bfa998684b94c0b3248d7bdad"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structstat__t.html">stat_t</a> haploid_highd::get_diversity_statistics </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>n_sample</em> = <code>1000</code></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculate diversity in the current population (Hamming distance between pairs of sequences) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n_sample</td><td>size of the statistical sample to use (the whole pop is often too large)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>mean and variance of the diversity in a <a class="el" href="structstat__t.html" title="Structure for short summary statistics.">stat_t</a> </dd></dl>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_abf5d073bfa998684b94c0b3248d7bdad_cgraph.png" border="0" usemap="#classhaploid__highd_abf5d073bfa998684b94c0b3248d7bdad_cgraph" alt=""/></div>
<map name="classhaploid__highd_abf5d073bfa998684b94c0b3248d7bdad_cgraph" id="classhaploid__highd_abf5d073bfa998684b94c0b3248d7bdad_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ab3e442c79f83e603bc30892e2a36e329" title="Produce and store a random sample of the population for stochastic processes." alt="" coords="597,5,752,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#aefe07878a3133687465c41804043779d" title="Sample random individuals from the population." alt="" coords="193,34,345,75"/><area shape="rect" id="node10" href="classhaploid__highd.html#a686aa6ae9c4d31f96dd6bb3d83d11188" title="haploid_highd::distance\l_Hamming" alt="" coords="190,123,349,165"/><area shape="rect" id="node7" href="classhaploid__highd.html#ab64531c02f245f39d9446efb5fcabe01" title="Get a random clone from the population." alt="" coords="397,34,549,75"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ac6029079cbb7ff559ed75b581ff8deef"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_fitness </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a2dae66421820a2248db9be1771ce9e62"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::get_fitness_histogram </td>
          <td>(</td>
          <td class="paramtype">gsl_histogram **&#160;</td>
          <td class="paramname"><em>hist</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>bins</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>n_sample</em> = <code>1000</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculate histogram of fitness from traits. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">hist</td><td>pointer to the gsl_histogram to fill </td></tr>
    <tr><td class="paramname">bins</td><td>number of bins in the histogram</td></tr>
  </table>
  </dd>
</dl>
<p><em>Note</em>: the output histogram might have less bins than requested if the sample size is too small.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n_sample</td><td>size of the random sample to use (the whole population is often too large)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>zero if successful, error codes otherwise</dd></dl>
<p>There is a small problem here, namely that sometimes the fitness distribution has a horrible tail which messes up the calculation of the bin width. Thus we first calculate the fitness average and variance, and then set the bin width so that the deleterious tail is within 2 sigma or so.</p>
<p><em>Note</em>: this function allocates memory for the histogram <em>only</em> if there is no binning error. If you get HP_NOBINSERR, no memory is allocated. The user is expected to release the memory manually. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a2dae66421820a2248db9be1771ce9e62_cgraph.png" border="0" usemap="#classhaploid__highd_a2dae66421820a2248db9be1771ce9e62_cgraph" alt=""/></div>
<map name="classhaploid__highd_a2dae66421820a2248db9be1771ce9e62_cgraph" id="classhaploid__highd_a2dae66421820a2248db9be1771ce9e62_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ab3e442c79f83e603bc30892e2a36e329" title="Produce and store a random sample of the population for stochastic processes." alt="" coords="584,5,739,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#aefe07878a3133687465c41804043779d" title="Sample random individuals from the population." alt="" coords="184,34,336,75"/><area shape="rect" id="node10" href="classhaploid__highd.html#a6dd67288b3fc216d2b83d7e4861966af" title="Calculate and store fitness population statistics." alt="" coords="193,99,327,141"/><area shape="rect" id="node7" href="classhaploid__highd.html#ab64531c02f245f39d9446efb5fcabe01" title="Get a random clone from the population." alt="" coords="384,34,536,75"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a4271fb90101fdea28032029efffbd395"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structstat__t.html">stat_t</a> haploid_highd::get_fitness_statistics </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a3a2a61046b0810e5b7b3794b49434aad"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::get_generation </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a2f6a793822b830dbb67fdb2d22c22194"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">string haploid_highd::get_genotype_string </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a99d356f0eac94d6710f54b8a7f303f4a"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_LD </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>locus1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>locus2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a0aaa82167545768f814c7e28868eee77"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_logmean_expfitness </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get the population exp-average of fitness, used for keeping the population size fixed. </p>
<dl class="section return"><dt>Returns</dt><dd>the population exp-average of fitness</dd></dl>
<p>Mathematically, this is \( \log\left( \left&lt; e^(F-F_{max}) \right&gt; \right) \). The baseline is \( F_{max} \) in order to avoid exponentiating large numbers. </p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a0aaa82167545768f814c7e28868eee77_icgraph.png" border="0" usemap="#classhaploid__highd_a0aaa82167545768f814c7e28868eee77_icgraph" alt=""/></div>
<map name="classhaploid__highd_a0aaa82167545768f814c7e28868eee77_icgraph" id="classhaploid__highd_a0aaa82167545768f814c7e28868eee77_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ae64df329db764e98e77398a775a43600" title="Get the log of the exp&#45;average fitness plus relaxation term." alt="" coords="197,5,363,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#aec2c31155f262a9c1cf21262eea2d03a" title="Generate offspring according to fitness (selection) and segregate some for sexual mating..." alt="" coords="411,5,555,46"/><area shape="rect" id="node7" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="603,12,746,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a7d00bee54a9dcddcec55d873545047cb"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_max_fitness </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ad4c1f8246b0c6ae14745d9634ff04ee6"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_moment </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>locus1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>locus2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a6e915dc6199d83ba2e35b3195b5b03b9"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::get_number_of_clones </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a6e915dc6199d83ba2e35b3195b5b03b9_icgraph.png" border="0" usemap="#classhaploid__highd_a6e915dc6199d83ba2e35b3195b5b03b9_icgraph" alt=""/></div>
<map name="classhaploid__highd_a6e915dc6199d83ba2e35b3195b5b03b9_icgraph" id="classhaploid__highd_a6e915dc6199d83ba2e35b3195b5b03b9_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af9d9998d0510c5fde8d3a422f900d731" title="Calculate the cumulative partition of sequences into clones." alt="" coords="189,5,344,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a303677b163c6b7edbf5f39a63ab84693"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::get_number_of_loci </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a303677b163c6b7edbf5f39a63ab84693_icgraph.png" border="0" usemap="#classhaploid__highd_a303677b163c6b7edbf5f39a63ab84693_icgraph" alt=""/></div>
<map name="classhaploid__highd_a303677b163c6b7edbf5f39a63ab84693_icgraph" id="classhaploid__highd_a303677b163c6b7edbf5f39a63ab84693_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="182,5,325,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a9e28c79bc0f9f68d705e0bf1643481f8"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::get_number_of_traits </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a061072acad96faee44962a5800f78281"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt; double &gt; haploid_highd::get_pair_frequencies </td>
          <td>(</td>
          <td class="paramtype">vector&lt; vector&lt; int &gt; &gt; *&#160;</td>
          <td class="paramname"><em>loci</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the joint frequency of two alleles, for a vector of allele pairs. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">loci</td><td>pointer to a vector of allele pairs. Each element of loci must be a vector of length 2.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>vector of joint frequencies. </dd></dl>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a061072acad96faee44962a5800f78281_cgraph.png" border="0" usemap="#classhaploid__highd_a061072acad96faee44962a5800f78281_cgraph" alt=""/></div>
<map name="classhaploid__highd_a061072acad96faee44962a5800f78281_cgraph" id="classhaploid__highd_a061072acad96faee44962a5800f78281_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a03dee5a44b1ca487f2e7dd72dea3aabd" title="Get the joint frequency of two alleles." alt="" coords="182,5,309,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a03dee5a44b1ca487f2e7dd72dea3aabd"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_pair_frequency </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>locus1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>locus2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the joint frequency of two alleles. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">locus1</td><td>position of the first allele. </td></tr>
    <tr><td class="paramname">locus2</td><td>position of the second allele.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the joint frequency of the two alleles. </dd></dl>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a03dee5a44b1ca487f2e7dd72dea3aabd_icgraph.png" border="0" usemap="#classhaploid__highd_a03dee5a44b1ca487f2e7dd72dea3aabd_icgraph" alt=""/></div>
<map name="classhaploid__highd_a03dee5a44b1ca487f2e7dd72dea3aabd_icgraph" id="classhaploid__highd_a03dee5a44b1ca487f2e7dd72dea3aabd_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a061072acad96faee44962a5800f78281" title="Get the joint frequency of two alleles, for a vector of allele pairs." alt="" coords="182,5,309,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a2e29536be4ec4c75d2889e5a2a28642e"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_participation_ratio </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a5f2fe58daa26fce5a3cea94adf92fe97"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::get_population_size </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a5f2fe58daa26fce5a3cea94adf92fe97_icgraph.png" border="0" usemap="#classhaploid__highd_a5f2fe58daa26fce5a3cea94adf92fe97_icgraph" alt=""/></div>
<map name="classhaploid__highd_a5f2fe58daa26fce5a3cea94adf92fe97_icgraph" id="classhaploid__highd_a5f2fe58daa26fce5a3cea94adf92fe97_icgraph">
<area shape="rect" id="node3" href="classhivpopulation.html#a53823e479fe82c35986fb9a932a1e576" title="hivpopulation::write\l_genotypes" alt="" coords="181,5,312,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a643f680f350c9060df4b01cd856cc50e"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::get_random_seed </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get a random seed from /dev/urandom. </p>
<dl class="section return"><dt>Returns</dt><dd>non-deterministic, random seed </dd></dl>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a643f680f350c9060df4b01cd856cc50e_icgraph.png" border="0" usemap="#classhaploid__highd_a643f680f350c9060df4b01cd856cc50e_icgraph" alt=""/></div>
<map name="classhaploid__highd_a643f680f350c9060df4b01cd856cc50e_icgraph" id="classhaploid__highd_a643f680f350c9060df4b01cd856cc50e_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ae58c4326bc251e8a86d3885f3297a25e" title="Default constructor." alt="" coords="182,5,333,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a1da3a90a51c42ccccbd9c47a93817052"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_trait </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>t</em> = <code>0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a5c6adc6e23625ec7eda23253074b564c"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_trait_covariance </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>t1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>t2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ac4d5cd6a0e86ae30257f41c1e5b889c5"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::get_trait_difference </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structclone__t.html">clone_t</a> *&#160;</td>
          <td class="paramname"><em>tempgt1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structclone__t.html">clone_t</a> *&#160;</td>
          <td class="paramname"><em>tempgt2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; int &gt; &amp;&#160;</td>
          <td class="paramname"><em>diffpos</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>traitnum</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Calculate trait difference between two clones. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">tempgt1</td><td>first clone </td></tr>
    <tr><td class="paramname">tempgt2</td><td>second clone </td></tr>
    <tr><td class="paramname">diffpos</td><td>positions at which the genotypes differ </td></tr>
    <tr><td class="paramname">traitnum</td><td>number of the trait to calculate</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>vector of differences</dd></dl>
<p>The operation is, for each trait, tempgt1 - tempgt2. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ac4d5cd6a0e86ae30257f41c1e5b889c5_cgraph.png" border="0" usemap="#classhaploid__highd_ac4d5cd6a0e86ae30257f41c1e5b889c5_cgraph" alt=""/></div>
<map name="classhaploid__highd_ac4d5cd6a0e86ae30257f41c1e5b889c5_cgraph" id="classhaploid__highd_ac4d5cd6a0e86ae30257f41c1e5b889c5_cgraph">
<area shape="rect" id="node3" href="classhypercube__highd.html#a8352b2198e0b8f6d352face30dff6989" title="Calculate difference between two hypercube points efficiently." alt="" coords="181,5,328,46"/><area shape="rect" id="node5" href="classhypercube__highd.html#a44781cd1dcf006cd33263ca5fbdf245a" title=": get the trait coefficient of a locus." alt="" coords="376,5,523,46"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ac4d5cd6a0e86ae30257f41c1e5b889c5_icgraph.png" border="0" usemap="#classhaploid__highd_ac4d5cd6a0e86ae30257f41c1e5b889c5_icgraph" alt=""/></div>
<map name="classhaploid__highd_ac4d5cd6a0e86ae30257f41c1e5b889c5_icgraph" id="classhaploid__highd_ac4d5cd6a0e86ae30257f41c1e5b889c5_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a943354008a9931e4223f952f4790efd6" title="Flip a spin (locus) in individual." alt="" coords="181,27,307,69"/><area shape="rect" id="node5" href="classhaploid__highd.html#af173582ee41aecd03d8f97aa234cf879" title="Mutate random clones at random loci (all loci)" alt="" coords="355,5,504,32"/><area shape="rect" id="node9" href="classhaploid__highd.html#a710a61fe8975aaaf4d148b2da20985ae" title="Flip a spin at a specific locus in random individual." alt="" coords="367,57,492,98"/><area shape="rect" id="node7" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="553,5,695,32"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a3554acf03e30ec4df69de98c29e1b2a8"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structstat__t.html">stat_t</a> haploid_highd::get_trait_statistics </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>t</em> = <code>0</code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a246641ceeb4e3a8a7168426fc2bba212"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::L </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="af173582ee41aecd03d8f97aa234cf879"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::mutate </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Mutate random clones at random loci (all loci) </p>
<p>Note: for efficiency reasons, if the mutation rate is zero, an if cycle skips the body altogether. The user can therefore call this function safely even if in non-mutating populations, without loss of performance. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_af173582ee41aecd03d8f97aa234cf879_cgraph.png" border="0" usemap="#classhaploid__highd_af173582ee41aecd03d8f97aa234cf879_cgraph" alt=""/></div>
<map name="classhaploid__highd_af173582ee41aecd03d8f97aa234cf879_cgraph" id="classhaploid__highd_af173582ee41aecd03d8f97aa234cf879_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ab3e442c79f83e603bc30892e2a36e329" title="Produce and store a random sample of the population for stochastic processes." alt="" coords="417,5,572,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#a943354008a9931e4223f952f4790efd6" title="Flip a spin (locus) in individual." alt="" coords="216,135,341,177"/><area shape="rect" id="node17" href="classhaploid__highd.html#ab64531c02f245f39d9446efb5fcabe01" title="Get a random clone from the population." alt="" coords="203,63,355,105"/><area shape="rect" id="node7" href="classhaploid__highd.html#ac4d5cd6a0e86ae30257f41c1e5b889c5" title="Calculate trait difference between two clones." alt="" coords="431,70,558,111"/><area shape="rect" id="node13" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="428,135,561,191"/><area shape="rect" id="node15" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="403,214,586,255"/><area shape="rect" id="node9" href="classhypercube__highd.html#a8352b2198e0b8f6d352face30dff6989" title="Calculate difference between two hypercube points efficiently." alt="" coords="635,70,781,111"/><area shape="rect" id="node11" href="classhypercube__highd.html#a44781cd1dcf006cd33263ca5fbdf245a" title=": get the trait coefficient of a locus." alt="" coords="829,70,976,111"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_af173582ee41aecd03d8f97aa234cf879_icgraph.png" border="0" usemap="#classhaploid__highd_af173582ee41aecd03d8f97aa234cf879_icgraph" alt=""/></div>
<map name="classhaploid__highd_af173582ee41aecd03d8f97aa234cf879_icgraph" id="classhaploid__highd_af173582ee41aecd03d8f97aa234cf879_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="203,5,346,32"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aaa0a39abeeb60600158b8430739280ec"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::N </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ab35ed279920654c5010561b692da200b"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::new_generation </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="af9d9998d0510c5fde8d3a422f900d731"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::partition_cumulative </td>
          <td>(</td>
          <td class="paramtype">vector&lt; unsigned int &gt; &amp;&#160;</td>
          <td class="paramname"><em>partition_cum</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Calculate the cumulative partition of sequences into clones. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">partition_cum</td><td>the vector to be filled</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>vector of cumulative clone sizes</dd></dl>
<p><em>Example</em>: if there are three clones of sizes (100, 22, 3) this function will fill the vector (100, 122, 125). The last element is of course the population size See also get_population_size.</p>
<p><em>Note</em>: the vector is taken in input by reference for performance reasons, since it can get huge. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_af9d9998d0510c5fde8d3a422f900d731_cgraph.png" border="0" usemap="#classhaploid__highd_af9d9998d0510c5fde8d3a422f900d731_cgraph" alt=""/></div>
<map name="classhaploid__highd_af9d9998d0510c5fde8d3a422f900d731_cgraph" id="classhaploid__highd_af9d9998d0510c5fde8d3a422f900d731_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a6e915dc6199d83ba2e35b3195b5b03b9" title="haploid_highd::get\l_number_of_clones" alt="" coords="209,5,343,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ae25ec86176f438db2dc0194f04ecfc42"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::print_allele_frequencies </td>
          <td>(</td>
          <td class="paramtype">ostream &amp;&#160;</td>
          <td class="paramname"><em>out</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print all allele frequencies into a stream provided. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">out</td><td>stream to put the allele frequencies (usually a file or stdout)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>zero if successful, nonzero otherwise </dd></dl>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ae25ec86176f438db2dc0194f04ecfc42_cgraph.png" border="0" usemap="#classhaploid__highd_ae25ec86176f438db2dc0194f04ecfc42_cgraph" alt=""/></div>
<map name="classhaploid__highd_ae25ec86176f438db2dc0194f04ecfc42_cgraph" id="classhaploid__highd_ae25ec86176f438db2dc0194f04ecfc42_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="189,149,352,176"/><area shape="rect" id="node5" href="classhaploid__highd.html#ad5eee5bc7f757e20fb3119bb01c59b1e" title="For each clone, recalculate its traits." alt="" coords="401,9,549,50"/><area shape="rect" id="node11" href="classhaploid__highd.html#a9b8ada9362e3108ee8994634231541be" title="For each clone, update fitness assuming traits are already up to date." alt="" coords="401,77,549,118"/><area shape="rect" id="node17" href="classhaploid__highd.html#afb3de45e613ffe03eadcc10f81c13448" title="Calculate and store trait population statistics and covariances." alt="" coords="408,142,541,183"/><area shape="rect" id="node19" href="classhaploid__highd.html#a6dd67288b3fc216d2b83d7e4861966af" title="Calculate and store fitness population statistics." alt="" coords="408,207,541,249"/><area shape="rect" id="node21" href="classhaploid__highd.html#a93cb185958cc78ce66c909d8df532a57" title="calculate and store allele frequencies" alt="" coords="408,273,541,314"/><area shape="rect" id="node7" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="623,5,756,46"/><area shape="rect" id="node9" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="830,12,1007,39"/><area shape="rect" id="node13" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="623,69,756,125"/><area shape="rect" id="node15" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="598,149,781,190"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ab3e442c79f83e603bc30892e2a36e329"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::produce_random_sample </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>size</em> = <code>1000</code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Produce and store a random sample of the population for stochastic processes. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">size</td><td>size of the sample</td></tr>
  </table>
  </dd>
</dl>
<p>Produce a random sample of genotypes (labeled by their clone of origin) which is to be used for mutations and sampling. The vector random_sample exists for efficiency reasons, for it is faster to call the random number generator once than many times.</p>
<p>Note: if you want to get random clones, please use <a class="el" href="classhaploid__highd.html#ab64531c02f245f39d9446efb5fcabe01" title="Get a random clone from the population.">random_clone()</a>. </p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ab3e442c79f83e603bc30892e2a36e329_icgraph.png" border="0" usemap="#classhaploid__highd_ab3e442c79f83e603bc30892e2a36e329_icgraph" alt=""/></div>
<map name="classhaploid__highd_ab3e442c79f83e603bc30892e2a36e329_icgraph" id="classhaploid__highd_ab3e442c79f83e603bc30892e2a36e329_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af173582ee41aecd03d8f97aa234cf879" title="Mutate random clones at random loci (all loci)" alt="" coords="409,113,559,140"/><area shape="rect" id="node7" href="classhaploid__highd.html#ab64531c02f245f39d9446efb5fcabe01" title="Get a random clone from the population." alt="" coords="208,230,360,271"/><area shape="rect" id="node14" href="classhaploid__highd.html#af77958954fe6001b2640b70bcc3681a4" title="Calculate mean and variance of the divergence from the [00...0] bitset." alt="" coords="610,5,761,46"/><area shape="rect" id="node16" href="classhaploid__highd.html#abf5d073bfa998684b94c0b3248d7bdad" title="Calculate diversity in the current population (Hamming distance between pairs of sequences)" alt="" coords="617,70,753,111"/><area shape="rect" id="node18" href="classhaploid__highd.html#a2dae66421820a2248db9be1771ce9e62" title="Calculate histogram of fitness from traits." alt="" coords="620,205,751,246"/><area shape="rect" id="node20" href="classhaploid__highd.html#a234a418eff9319a2dfb6371b73220b56" title="Get histogram of divergence from the [00...0] bitset." alt="" coords="608,305,763,346"/><area shape="rect" id="node22" href="classhaploid__highd.html#a25a490ade017cff4cccb88ced04261ce" title="Get histogram of diversity in the population (mutual Hamming distance)" alt="" coords="615,438,755,479"/><area shape="rect" id="node24" href="classhivpopulation.html#a53823e479fe82c35986fb9a932a1e576" title="hivpopulation::write\l_genotypes" alt="" coords="419,295,549,337"/><area shape="rect" id="node5" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="614,136,757,163"/><area shape="rect" id="node10" href="classhaploid__highd.html#a710a61fe8975aaaf4d148b2da20985ae" title="Flip a spin at a specific locus in random individual." alt="" coords="421,230,547,271"/><area shape="rect" id="node12" href="classhaploid__highd.html#aefe07878a3133687465c41804043779d" title="Sample random individuals from the population." alt="" coords="408,165,560,206"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ab64531c02f245f39d9446efb5fcabe01"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::random_clone </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get a random clone from the population. </p>
<dl class="section return"><dt>Returns</dt><dd>the index of the random clone</dd></dl>
<p>The probability density function from which the individual is chosen is flat over the population (larger clones are proportionally more likely to be returned here).</p>
<p>Note: for efficiency reasons, the class keeps a storage of random clone indices. If you need much more than, say, 1000 random clones at a time, please call produce_random_sample with the required size in advance. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ab64531c02f245f39d9446efb5fcabe01_cgraph.png" border="0" usemap="#classhaploid__highd_ab64531c02f245f39d9446efb5fcabe01_cgraph" alt=""/></div>
<map name="classhaploid__highd_ab64531c02f245f39d9446efb5fcabe01_cgraph" id="classhaploid__highd_ab64531c02f245f39d9446efb5fcabe01_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ab3e442c79f83e603bc30892e2a36e329" title="Produce and store a random sample of the population for stochastic processes." alt="" coords="205,5,360,46"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ab64531c02f245f39d9446efb5fcabe01_icgraph.png" border="0" usemap="#classhaploid__highd_ab64531c02f245f39d9446efb5fcabe01_icgraph" alt=""/></div>
<map name="classhaploid__highd_ab64531c02f245f39d9446efb5fcabe01_icgraph" id="classhaploid__highd_ab64531c02f245f39d9446efb5fcabe01_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af173582ee41aecd03d8f97aa234cf879" title="Mutate random clones at random loci (all loci)" alt="" coords="207,23,356,49"/><area shape="rect" id="node7" href="classhaploid__highd.html#a710a61fe8975aaaf4d148b2da20985ae" title="Flip a spin at a specific locus in random individual." alt="" coords="219,106,344,147"/><area shape="rect" id="node9" href="classhaploid__highd.html#aefe07878a3133687465c41804043779d" title="Sample random individuals from the population." alt="" coords="205,171,357,213"/><area shape="rect" id="node21" href="classhivpopulation.html#a53823e479fe82c35986fb9a932a1e576" title="hivpopulation::write\l_genotypes" alt="" coords="216,237,347,278"/><area shape="rect" id="node5" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="411,5,554,32"/><area shape="rect" id="node11" href="classhaploid__highd.html#af77958954fe6001b2640b70bcc3681a4" title="Calculate mean and variance of the divergence from the [00...0] bitset." alt="" coords="407,57,558,98"/><area shape="rect" id="node13" href="classhaploid__highd.html#abf5d073bfa998684b94c0b3248d7bdad" title="Calculate diversity in the current population (Hamming distance between pairs of sequences)" alt="" coords="415,122,551,163"/><area shape="rect" id="node15" href="classhaploid__highd.html#a2dae66421820a2248db9be1771ce9e62" title="Calculate histogram of fitness from traits." alt="" coords="417,187,548,229"/><area shape="rect" id="node17" href="classhaploid__highd.html#a234a418eff9319a2dfb6371b73220b56" title="Get histogram of divergence from the [00...0] bitset." alt="" coords="405,253,560,294"/><area shape="rect" id="node19" href="classhaploid__highd.html#a25a490ade017cff4cccb88ced04261ce" title="Get histogram of diversity in the population (mutual Hamming distance)" alt="" coords="413,318,553,359"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aefe07878a3133687465c41804043779d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::random_clones </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>n_o_individuals</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; int &gt; *&#160;</td>
          <td class="paramname"><em>sample</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Sample random individuals from the population. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n_o_individuals</td><td>number of individuals to sample </td></tr>
    <tr><td class="paramname">sample</td><td>pointer to vector where to put the result</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>zero if successful, nonzero otherwise</dd></dl>
<p>The results are not returned as a vector for performance reasons, as one might want to get a lot of random clones. *sample may be not empty (but must be allocated). In any case, clone numbers of the sampled individuals are appended to *sample. Hence, you can use this function iteratively (although there might not be a good reason to do so). </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_aefe07878a3133687465c41804043779d_cgraph.png" border="0" usemap="#classhaploid__highd_aefe07878a3133687465c41804043779d_cgraph" alt=""/></div>
<map name="classhaploid__highd_aefe07878a3133687465c41804043779d_cgraph" id="classhaploid__highd_aefe07878a3133687465c41804043779d_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ab64531c02f245f39d9446efb5fcabe01" title="Get a random clone from the population." alt="" coords="205,5,357,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#ab3e442c79f83e603bc30892e2a36e329" title="Produce and store a random sample of the population for stochastic processes." alt="" coords="405,5,560,46"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_aefe07878a3133687465c41804043779d_icgraph.png" border="0" usemap="#classhaploid__highd_aefe07878a3133687465c41804043779d_icgraph" alt=""/></div>
<map name="classhaploid__highd_aefe07878a3133687465c41804043779d_icgraph" id="classhaploid__highd_aefe07878a3133687465c41804043779d_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af77958954fe6001b2640b70bcc3681a4" title="Calculate mean and variance of the divergence from the [00...0] bitset." alt="" coords="207,5,358,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#abf5d073bfa998684b94c0b3248d7bdad" title="Calculate diversity in the current population (Hamming distance between pairs of sequences)" alt="" coords="215,70,351,111"/><area shape="rect" id="node7" href="classhaploid__highd.html#a2dae66421820a2248db9be1771ce9e62" title="Calculate histogram of fitness from traits." alt="" coords="217,135,348,177"/><area shape="rect" id="node9" href="classhaploid__highd.html#a234a418eff9319a2dfb6371b73220b56" title="Get histogram of divergence from the [00...0] bitset." alt="" coords="205,201,360,242"/><area shape="rect" id="node11" href="classhaploid__highd.html#a25a490ade017cff4cccb88ced04261ce" title="Get histogram of diversity in the population (mutual Hamming distance)" alt="" coords="213,266,353,307"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a60617675023cab02a599a1ea7f9605c6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::read_ms_sample </td>
          <td>(</td>
          <td class="paramtype">istream &amp;&#160;</td>
          <td class="paramname"><em>gts</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>skip_locus</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>multiplicity</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Read the output of Hudson's ms and use it to initialize the genotype distribution. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">gts</td><td>genotypes output of <em>ms</em> </td></tr>
    <tr><td class="paramname">skip_locus</td><td>position of the locus to be skipped </td></tr>
    <tr><td class="paramname">multiplicity</td><td>number of times each genotype is added</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>zero if successful, error codes otherwise</dd></dl>
<p>ms loci are fed into the genotype with a locus that is skipped. Each ms genotype is added multiple times. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a60617675023cab02a599a1ea7f9605c6_cgraph.png" border="0" usemap="#classhaploid__highd_a60617675023cab02a599a1ea7f9605c6_cgraph" alt=""/></div>
<map name="classhaploid__highd_a60617675023cab02a599a1ea7f9605c6_cgraph" id="classhaploid__highd_a60617675023cab02a599a1ea7f9605c6_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4" title="Initialize a wildtype population (00...0)" alt="" coords="190,67,317,109"/><area shape="rect" id="node5" href="classhaploid__highd.html#a44235ff9f17f4e81f290611d7829818a" title="Add the genotype specified by a bitset to the current population in in n copies." alt="" coords="585,9,716,50"/><area shape="rect" id="node15" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="365,180,528,207"/><area shape="rect" id="node7" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="799,6,932,47"/><area shape="rect" id="node11" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="799,136,932,192"/><area shape="rect" id="node13" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="774,71,957,113"/><area shape="rect" id="node9" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="1006,13,1183,40"/><area shape="rect" id="node17" href="classhaploid__highd.html#ad5eee5bc7f757e20fb3119bb01c59b1e" title="For each clone, recalculate its traits." alt="" coords="577,74,725,115"/><area shape="rect" id="node20" href="classhaploid__highd.html#a9b8ada9362e3108ee8994634231541be" title="For each clone, update fitness assuming traits are already up to date." alt="" coords="577,139,725,181"/><area shape="rect" id="node24" href="classhaploid__highd.html#afb3de45e613ffe03eadcc10f81c13448" title="Calculate and store trait population statistics and covariances." alt="" coords="584,205,717,246"/><area shape="rect" id="node26" href="classhaploid__highd.html#a6dd67288b3fc216d2b83d7e4861966af" title="Calculate and store fitness population statistics." alt="" coords="584,270,717,311"/><area shape="rect" id="node28" href="classhaploid__highd.html#a93cb185958cc78ce66c909d8df532a57" title="calculate and store allele frequencies" alt="" coords="584,335,717,377"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a0e08d1847e521402cf0e2676d6ed38b9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::read_ms_sample_sparse </td>
          <td>(</td>
          <td class="paramtype">istream &amp;&#160;</td>
          <td class="paramname"><em>gts</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>skip_locus</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>multiplicity</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>distance</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Read the output of Hudson's ms and use it to initialize the genotype distribution. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">gts</td><td>genotypes output of <em>ms</em> </td></tr>
    <tr><td class="paramname">skip_locus</td><td>positionof the locus to be skipped </td></tr>
    <tr><td class="paramname">multiplicity</td><td>number of times each genotype must be added </td></tr>
    <tr><td class="paramname">distance</td><td>distance between two polymorphic sites</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>zero if successful, error codes otherwise</dd></dl>
<p>ms loci are fed into the genotype at distance "distance", i.e. there are distance-1 monomorphic loci. One locus is skipped. Each ms genotype is added multiple times. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a0e08d1847e521402cf0e2676d6ed38b9_cgraph.png" border="0" usemap="#classhaploid__highd_a0e08d1847e521402cf0e2676d6ed38b9_cgraph" alt=""/></div>
<map name="classhaploid__highd_a0e08d1847e521402cf0e2676d6ed38b9_cgraph" id="classhaploid__highd_a0e08d1847e521402cf0e2676d6ed38b9_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4" title="Initialize a wildtype population (00...0)" alt="" coords="198,66,325,107"/><area shape="rect" id="node5" href="classhaploid__highd.html#a44235ff9f17f4e81f290611d7829818a" title="Add the genotype specified by a bitset to the current population in in n copies." alt="" coords="593,7,724,49"/><area shape="rect" id="node30" href="classhaploid__highd.html#a303677b163c6b7edbf5f39a63ab84693" title="haploid_highd::get\l_number_of_loci" alt="" coords="198,131,325,173"/><area shape="rect" id="node15" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="373,179,536,205"/><area shape="rect" id="node7" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="807,5,940,46"/><area shape="rect" id="node11" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="807,135,940,191"/><area shape="rect" id="node13" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="782,70,965,111"/><area shape="rect" id="node9" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="1014,12,1191,39"/><area shape="rect" id="node17" href="classhaploid__highd.html#ad5eee5bc7f757e20fb3119bb01c59b1e" title="For each clone, recalculate its traits." alt="" coords="585,73,733,114"/><area shape="rect" id="node20" href="classhaploid__highd.html#a9b8ada9362e3108ee8994634231541be" title="For each clone, update fitness assuming traits are already up to date." alt="" coords="585,138,733,179"/><area shape="rect" id="node24" href="classhaploid__highd.html#afb3de45e613ffe03eadcc10f81c13448" title="Calculate and store trait population statistics and covariances." alt="" coords="592,203,725,245"/><area shape="rect" id="node26" href="classhaploid__highd.html#a6dd67288b3fc216d2b83d7e4861966af" title="Calculate and store fitness population statistics." alt="" coords="592,269,725,310"/><area shape="rect" id="node28" href="classhaploid__highd.html#a93cb185958cc78ce66c909d8df532a57" title="calculate and store allele frequencies" alt="" coords="592,334,725,375"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ac79972f0e845b52b3191a27ddae309a8"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">boost::dynamic_bitset haploid_highd::reassortment_pattern </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Produce a random reassortement pattern. </p>
<dl class="section return"><dt>Returns</dt><dd>reassortement pattern </dd></dl>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ac79972f0e845b52b3191a27ddae309a8_icgraph.png" border="0" usemap="#classhaploid__highd_ac79972f0e845b52b3191a27ddae309a8_icgraph" alt=""/></div>
<map name="classhaploid__highd_ac79972f0e845b52b3191a27ddae309a8_icgraph" id="classhaploid__highd_ac79972f0e845b52b3191a27ddae309a8_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#afb3d9e695f16b8412902757f18cfb63e" title="Recombine two genotypes parent1 and parent2 to produce two new genotypes." alt="" coords="241,12,410,39"/><area shape="rect" id="node5" href="classhaploid__highd.html#ab08783cb783ca66ff2c87227dbd74314" title="Pair and mate sexual gametes." alt="" coords="459,5,589,46"/><area shape="rect" id="node7" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="638,12,781,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="afb3d9e695f16b8412902757f18cfb63e"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::recombine </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>parent1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>parent2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Recombine two genotypes parent1 and parent2 to produce two new genotypes. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">parent1</td><td>first parent </td></tr>
    <tr><td class="paramname">parent2</td><td>second parent</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>zero if successful</dd></dl>
<p>The new genotypes are stored in new_pop at positions ng and ng+1. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_afb3d9e695f16b8412902757f18cfb63e_cgraph.png" border="0" usemap="#classhaploid__highd_afb3d9e695f16b8412902757f18cfb63e_cgraph" alt=""/></div>
<map name="classhaploid__highd_afb3d9e695f16b8412902757f18cfb63e_cgraph" id="classhaploid__highd_afb3d9e695f16b8412902757f18cfb63e_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ac79972f0e845b52b3191a27ddae309a8" title="Produce a random reassortement pattern." alt="" coords="225,5,410,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#a33a62c8a5c57d36e071099906131eb31" title="Choose a number of crossover points and produce a crossover pattern." alt="" coords="235,70,399,111"/><area shape="rect" id="node7" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="251,135,384,177"/><area shape="rect" id="node11" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="251,200,384,256"/><area shape="rect" id="node13" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="226,279,409,321"/><area shape="rect" id="node9" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="459,143,637,169"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_afb3d9e695f16b8412902757f18cfb63e_icgraph.png" border="0" usemap="#classhaploid__highd_afb3d9e695f16b8412902757f18cfb63e_icgraph" alt=""/></div>
<map name="classhaploid__highd_afb3d9e695f16b8412902757f18cfb63e_icgraph" id="classhaploid__highd_afb3d9e695f16b8412902757f18cfb63e_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ab08783cb783ca66ff2c87227dbd74314" title="Pair and mate sexual gametes." alt="" coords="224,5,355,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="403,12,546,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ae9c993847dc49cd24811df7ef1fb9a5e"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::recombine_crossover </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>parent1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>parent2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>ng</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ae64df329db764e98e77398a775a43600"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::relaxation_value </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get the log of the exp-average fitness plus relaxation term. </p>
<dl class="section return"><dt>Returns</dt><dd>baseline relative to which growth rates are measured, i.e. the current mean fitness plus the term that causes the population size to relax to carrying_capacity </dd></dl>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ae64df329db764e98e77398a775a43600_cgraph.png" border="0" usemap="#classhaploid__highd_ae64df329db764e98e77398a775a43600_cgraph" alt=""/></div>
<map name="classhaploid__highd_ae64df329db764e98e77398a775a43600_cgraph" id="classhaploid__highd_ae64df329db764e98e77398a775a43600_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a0aaa82167545768f814c7e28868eee77" title="Get the population exp&#45;average of fitness, used for keeping the population size fixed." alt="" coords="219,5,363,46"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ae64df329db764e98e77398a775a43600_icgraph.png" border="0" usemap="#classhaploid__highd_ae64df329db764e98e77398a775a43600_icgraph" alt=""/></div>
<map name="classhaploid__highd_ae64df329db764e98e77398a775a43600_icgraph" id="classhaploid__highd_ae64df329db764e98e77398a775a43600_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#aec2c31155f262a9c1cf21262eea2d03a" title="Generate offspring according to fitness (selection) and segregate some for sexual mating..." alt="" coords="219,5,363,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="411,12,554,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aec2c31155f262a9c1cf21262eea2d03a"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::select_gametes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Generate offspring according to fitness (selection) and segregate some for sexual mating. </p>
<dl class="section return"><dt>Returns</dt><dd>zero if successful, error codes otherwise</dd></dl>
<p>Random Poisson offspring numbers are drawn from all parents, proportionally to their fitness. A fraction r of all offspring are designated for sexual reproduction, while the remainder (1-r) reproduces by exact duplication (mutations are introduced later). The population size relaxes to a carrying capacity, i.e. selection is soft but the population size is not exactly fixed.</p>
<p>TODO: extend sex_gametes to list of indices and old AND new population to spot redundant recombination events </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_aec2c31155f262a9c1cf21262eea2d03a_cgraph.png" border="0" usemap="#classhaploid__highd_aec2c31155f262a9c1cf21262eea2d03a_cgraph" alt=""/></div>
<map name="classhaploid__highd_aec2c31155f262a9c1cf21262eea2d03a_cgraph" id="classhaploid__highd_aec2c31155f262a9c1cf21262eea2d03a_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#ae64df329db764e98e77398a775a43600" title="Get the log of the exp&#45;average fitness plus relaxation term." alt="" coords="197,5,363,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#a0aaa82167545768f814c7e28868eee77" title="Get the population exp&#45;average of fitness, used for keeping the population size fixed." alt="" coords="411,5,555,46"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_aec2c31155f262a9c1cf21262eea2d03a_icgraph.png" border="0" usemap="#classhaploid__highd_aec2c31155f262a9c1cf21262eea2d03a_icgraph" alt=""/></div>
<map name="classhaploid__highd_aec2c31155f262a9c1cf21262eea2d03a_icgraph" id="classhaploid__highd_aec2c31155f262a9c1cf21262eea2d03a_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="198,12,341,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a3351949a49ad6702beb996b58755ce4e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::set_allele_frequencies </td>
          <td>(</td>
          <td class="paramtype">double *&#160;</td>
          <td class="paramname"><em>freq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>N_in</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Initialize population in linkage equilibrium. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">freq</td><td>target allele frequencies </td></tr>
    <tr><td class="paramname">N_in</td><td>number of individuals to be created</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>zero if successful, error codes otherwise</dd></dl>
<p><em>Note</em>: the carrying capacity is set equal to N_in if it is still unset. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a3351949a49ad6702beb996b58755ce4e_cgraph.png" border="0" usemap="#classhaploid__highd_a3351949a49ad6702beb996b58755ce4e_cgraph" alt=""/></div>
<map name="classhaploid__highd_a3351949a49ad6702beb996b58755ce4e_cgraph" id="classhaploid__highd_a3351949a49ad6702beb996b58755ce4e_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a44235ff9f17f4e81f290611d7829818a" title="Add the genotype specified by a bitset to the current population in in n copies." alt="" coords="407,70,537,111"/><area shape="rect" id="node13" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="187,188,349,215"/><area shape="rect" id="node5" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="620,38,753,79"/><area shape="rect" id="node9" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="620,132,753,188"/><area shape="rect" id="node11" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="595,270,778,311"/><area shape="rect" id="node7" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="827,45,1005,72"/><area shape="rect" id="node15" href="classhaploid__highd.html#ad5eee5bc7f757e20fb3119bb01c59b1e" title="For each clone, recalculate its traits." alt="" coords="398,5,546,46"/><area shape="rect" id="node18" href="classhaploid__highd.html#a9b8ada9362e3108ee8994634231541be" title="For each clone, update fitness assuming traits are already up to date." alt="" coords="398,331,546,373"/><area shape="rect" id="node22" href="classhaploid__highd.html#afb3de45e613ffe03eadcc10f81c13448" title="Calculate and store trait population statistics and covariances." alt="" coords="405,135,539,177"/><area shape="rect" id="node24" href="classhaploid__highd.html#a6dd67288b3fc216d2b83d7e4861966af" title="Calculate and store fitness population statistics." alt="" coords="405,201,539,242"/><area shape="rect" id="node26" href="classhaploid__highd.html#a93cb185958cc78ce66c909d8df532a57" title="calculate and store allele frequencies" alt="" coords="405,266,539,307"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aa0c34e02ed924603417fdb8109918f7e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::set_genotypes </td>
          <td>(</td>
          <td class="paramtype">vector&lt; <a class="el" href="structgenotype__value__pair__t.html">genotype_value_pair_t</a> &gt;&#160;</td>
          <td class="paramname"><em>gt</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Initialize the population with genotype counts. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">gt</td><td>vector of genotype_value_pair with genotypes and sizes</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>0 if successful, nonzero otherwise.</dd></dl>
<p><em>Note</em>: the population size is set as the total sum of counts of all genotypes. The carrying capacity is also set to the same number if it is still unset. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_aa0c34e02ed924603417fdb8109918f7e_cgraph.png" border="0" usemap="#classhaploid__highd_aa0c34e02ed924603417fdb8109918f7e_cgraph" alt=""/></div>
<map name="classhaploid__highd_aa0c34e02ed924603417fdb8109918f7e_cgraph" id="classhaploid__highd_aa0c34e02ed924603417fdb8109918f7e_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a44235ff9f17f4e81f290611d7829818a" title="Add the genotype specified by a bitset to the current population in in n copies." alt="" coords="401,70,532,111"/><area shape="rect" id="node13" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="181,188,344,215"/><area shape="rect" id="node5" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="615,38,748,79"/><area shape="rect" id="node9" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="615,132,748,188"/><area shape="rect" id="node11" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="590,270,773,311"/><area shape="rect" id="node7" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="822,45,999,72"/><area shape="rect" id="node15" href="classhaploid__highd.html#ad5eee5bc7f757e20fb3119bb01c59b1e" title="For each clone, recalculate its traits." alt="" coords="393,5,541,46"/><area shape="rect" id="node18" href="classhaploid__highd.html#a9b8ada9362e3108ee8994634231541be" title="For each clone, update fitness assuming traits are already up to date." alt="" coords="393,331,541,373"/><area shape="rect" id="node22" href="classhaploid__highd.html#afb3de45e613ffe03eadcc10f81c13448" title="Calculate and store trait population statistics and covariances." alt="" coords="400,135,533,177"/><area shape="rect" id="node24" href="classhaploid__highd.html#a6dd67288b3fc216d2b83d7e4861966af" title="Calculate and store fitness population statistics." alt="" coords="400,201,533,242"/><area shape="rect" id="node26" href="classhaploid__highd.html#a93cb185958cc78ce66c909d8df532a57" title="calculate and store allele frequencies" alt="" coords="400,266,533,307"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a0a7af3cbfabcbf0b3b01e4bcffd87425"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::set_random_epistasis </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>epistasis_std</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ae8d94a3616a3c1ffdb07844fbd0349aa"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::set_random_trait_epistasis </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>epistasis_std</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>traitnumber</em> = <code>0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a9418c16f238aacab3be2dbffb62246c4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::set_wildtype </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>N_in</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Initialize a wildtype population (00...0) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">N_in</td><td>number of individuals</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>0 if successful, nonzero otherwise.</dd></dl>
<p><em>Note</em>: the carrying capacity is set equal to N_in if it is still unset. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a9418c16f238aacab3be2dbffb62246c4_cgraph.png" border="0" usemap="#classhaploid__highd_a9418c16f238aacab3be2dbffb62246c4_cgraph" alt=""/></div>
<map name="classhaploid__highd_a9418c16f238aacab3be2dbffb62246c4_cgraph" id="classhaploid__highd_a9418c16f238aacab3be2dbffb62246c4_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a44235ff9f17f4e81f290611d7829818a" title="Add the genotype specified by a bitset to the current population in in n copies." alt="" coords="401,70,532,111"/><area shape="rect" id="node13" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="181,188,344,215"/><area shape="rect" id="node5" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="615,38,748,79"/><area shape="rect" id="node9" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="615,132,748,188"/><area shape="rect" id="node11" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="590,270,773,311"/><area shape="rect" id="node7" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="822,45,999,72"/><area shape="rect" id="node15" href="classhaploid__highd.html#ad5eee5bc7f757e20fb3119bb01c59b1e" title="For each clone, recalculate its traits." alt="" coords="393,5,541,46"/><area shape="rect" id="node18" href="classhaploid__highd.html#a9b8ada9362e3108ee8994634231541be" title="For each clone, update fitness assuming traits are already up to date." alt="" coords="393,331,541,373"/><area shape="rect" id="node22" href="classhaploid__highd.html#afb3de45e613ffe03eadcc10f81c13448" title="Calculate and store trait population statistics and covariances." alt="" coords="400,135,533,177"/><area shape="rect" id="node24" href="classhaploid__highd.html#a6dd67288b3fc216d2b83d7e4861966af" title="Calculate and store fitness population statistics." alt="" coords="400,201,533,242"/><area shape="rect" id="node26" href="classhaploid__highd.html#a93cb185958cc78ce66c909d8df532a57" title="calculate and store allele frequencies" alt="" coords="400,266,533,307"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a9418c16f238aacab3be2dbffb62246c4_icgraph.png" border="0" usemap="#classhaploid__highd_a9418c16f238aacab3be2dbffb62246c4_icgraph" alt=""/></div>
<map name="classhaploid__highd_a9418c16f238aacab3be2dbffb62246c4_icgraph" id="classhaploid__highd_a9418c16f238aacab3be2dbffb62246c4_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a60617675023cab02a599a1ea7f9605c6" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="203,5,338,46"/><area shape="rect" id="node5" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="199,70,342,111"/><area shape="rect" id="node7" href="classhivpopulation.html#a299863ae882da177d548edc26d9a2cb8" title="Construct a HIV population with certain parameters." alt="" coords="181,136,360,163"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a6c8674a3f1dcb55ea680b2fd7dbf655c"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::shuffle_genotypes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a6fa4ffe295eef314642dec0208f81b7d"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::swap_populations </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Make the the temporary population the current one. </p>
<p>After the new population is completely assembled, we have to make it the current population. This is implemented simply by swaping pointers. </p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a6fa4ffe295eef314642dec0208f81b7d_icgraph.png" border="0" usemap="#classhaploid__highd_a6fa4ffe295eef314642dec0208f81b7d_icgraph" alt=""/></div>
<map name="classhaploid__highd_a6fa4ffe295eef314642dec0208f81b7d_icgraph" id="classhaploid__highd_a6fa4ffe295eef314642dec0208f81b7d_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af7fbfc38495246eb5ad91201c9a8f58a" title="Evolve for some generations under the specified conditions." alt="" coords="206,5,349,32"/><area shape="rect" id="node5" href="classhaploid__highd.html#aae3666ef1b744f0a1918d5694ec5199b" title="Cause a bottleneck in the population size." alt="" coords="193,56,362,83"/><area shape="rect" id="node7" href="classhaploid__highd.html#a981292b7434288cbee87093576b61371" title="Remove duplicate clones." alt="" coords="204,107,351,149"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a981292b7434288cbee87093576b61371"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::unique_clones </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Remove duplicate clones. </p>
<p>The library does not usually check whether two clones with the same genotype are present, but this can happen in case of recurrent mutation. This function merges duplicates.</p>
<p><em>Note</em>: this is only needed for studying the clone structure. Evolution itself does not need to make sure that clones are unique. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a981292b7434288cbee87093576b61371_cgraph.png" border="0" usemap="#classhaploid__highd_a981292b7434288cbee87093576b61371_cgraph" alt=""/></div>
<map name="classhaploid__highd_a981292b7434288cbee87093576b61371_cgraph" id="classhaploid__highd_a981292b7434288cbee87093576b61371_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a6fa4ffe295eef314642dec0208f81b7d" title="Make the the temporary population the current one." alt="" coords="200,5,339,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a9b8ada9362e3108ee8994634231541be"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::update_fitness </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>For each clone, update fitness assuming traits are already up to date. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a9b8ada9362e3108ee8994634231541be_cgraph.png" border="0" usemap="#classhaploid__highd_a9b8ada9362e3108ee8994634231541be_cgraph" alt=""/></div>
<map name="classhaploid__highd_a9b8ada9362e3108ee8994634231541be_cgraph" id="classhaploid__highd_a9b8ada9362e3108ee8994634231541be_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#a7f51d6a4f5791961d1b262e15735f820" title="Calculate fitness from traits of the chosen clone." alt="" coords="228,5,361,61"/><area shape="rect" id="node5" href="classhaploid__highd.html#a0d929c061bbf1c2fd4d82f6bb2f2f998" title="haploid_highd::check\l_individual_maximal_fitness" alt="" coords="203,85,386,126"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_a9b8ada9362e3108ee8994634231541be_icgraph.png" border="0" usemap="#classhaploid__highd_a9b8ada9362e3108ee8994634231541be_icgraph" alt=""/></div>
<map name="classhaploid__highd_a9b8ada9362e3108ee8994634231541be_icgraph" id="classhaploid__highd_a9b8ada9362e3108ee8994634231541be_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="203,109,365,136"/><area shape="rect" id="node19" href="classhivpopulation.html#ad1caa476f2d93020d52d0ce46f81a529" title="hivpopulation::read\l_replication_coefficients" alt="" coords="204,161,364,202"/><area shape="rect" id="node21" href="classhivpopulation.html#a6afaadd90cf25378476a13a2bef80875" title="hivpopulation::read\l_resistance_coefficients" alt="" coords="203,226,365,267"/><area shape="rect" id="node23" href="classhivpopulation.html#a0f5bd29413e44f8a075afe164c05853e" title="hivpopulation::set\l_treatment" alt="" coords="223,291,345,333"/><area shape="rect" id="node5" href="classhaploid__highd.html#a3351949a49ad6702beb996b58755ce4e" title="Initialize population in linkage equilibrium." alt="" coords="415,5,547,46"/><area shape="rect" id="node7" href="classhaploid__highd.html#aa0c34e02ed924603417fdb8109918f7e" title="Initialize the population with genotype counts." alt="" coords="418,70,545,111"/><area shape="rect" id="node9" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4" title="Initialize a wildtype population (00...0)" alt="" coords="418,135,545,177"/><area shape="rect" id="node17" href="classhaploid__highd.html#ae25ec86176f438db2dc0194f04ecfc42" title="Print all allele frequencies into a stream provided." alt="" coords="414,201,549,242"/><area shape="rect" id="node11" href="classhaploid__highd.html#a60617675023cab02a599a1ea7f9605c6" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="619,70,754,111"/><area shape="rect" id="node13" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="615,135,758,177"/><area shape="rect" id="node15" href="classhivpopulation.html#a299863ae882da177d548edc26d9a2cb8" title="Construct a HIV population with certain parameters." alt="" coords="597,201,776,228"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ad5eee5bc7f757e20fb3119bb01c59b1e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void haploid_highd::update_traits </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>For each clone, recalculate its traits. </p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ad5eee5bc7f757e20fb3119bb01c59b1e_cgraph.png" border="0" usemap="#classhaploid__highd_ad5eee5bc7f757e20fb3119bb01c59b1e_cgraph" alt=""/></div>
<map name="classhaploid__highd_ad5eee5bc7f757e20fb3119bb01c59b1e_cgraph" id="classhaploid__highd_ad5eee5bc7f757e20fb3119bb01c59b1e_cgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#acf40ae60f1441f5a8bb67a5a84444c01" title="Calculate traits of the chosen clone." alt="" coords="203,5,336,46"/><area shape="rect" id="node5" href="classhypercube__highd.html#ab53deaa3338c2e2c597bd2b4431cd7b2" title="Get single value on the hypercube." alt="" coords="385,12,562,39"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classhaploid__highd_ad5eee5bc7f757e20fb3119bb01c59b1e_icgraph.png" border="0" usemap="#classhaploid__highd_ad5eee5bc7f757e20fb3119bb01c59b1e_icgraph" alt=""/></div>
<map name="classhaploid__highd_ad5eee5bc7f757e20fb3119bb01c59b1e_icgraph" id="classhaploid__highd_ad5eee5bc7f757e20fb3119bb01c59b1e_icgraph">
<area shape="rect" id="node3" href="classhaploid__highd.html#af40993ccd48558777ab65ab88bc590b3" title="Calculate trait and fitness statistics and allele frequences." alt="" coords="203,109,365,136"/><area shape="rect" id="node19" href="classhivpopulation.html#ad1caa476f2d93020d52d0ce46f81a529" title="hivpopulation::read\l_replication_coefficients" alt="" coords="204,161,364,202"/><area shape="rect" id="node21" href="classhivpopulation.html#a6afaadd90cf25378476a13a2bef80875" title="hivpopulation::read\l_resistance_coefficients" alt="" coords="203,226,365,267"/><area shape="rect" id="node23" href="classhivpopulation.html#a0f5bd29413e44f8a075afe164c05853e" title="hivpopulation::set\l_treatment" alt="" coords="223,291,345,333"/><area shape="rect" id="node5" href="classhaploid__highd.html#a3351949a49ad6702beb996b58755ce4e" title="Initialize population in linkage equilibrium." alt="" coords="415,5,547,46"/><area shape="rect" id="node7" href="classhaploid__highd.html#aa0c34e02ed924603417fdb8109918f7e" title="Initialize the population with genotype counts." alt="" coords="418,70,545,111"/><area shape="rect" id="node9" href="classhaploid__highd.html#a9418c16f238aacab3be2dbffb62246c4" title="Initialize a wildtype population (00...0)" alt="" coords="418,135,545,177"/><area shape="rect" id="node17" href="classhaploid__highd.html#ae25ec86176f438db2dc0194f04ecfc42" title="Print all allele frequencies into a stream provided." alt="" coords="414,201,549,242"/><area shape="rect" id="node11" href="classhaploid__highd.html#a60617675023cab02a599a1ea7f9605c6" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="619,70,754,111"/><area shape="rect" id="node13" href="classhaploid__highd.html#a0e08d1847e521402cf0e2676d6ed38b9" title="Read the output of Hudson&#39;s ms and use it to initialize the genotype distribution." alt="" coords="615,135,758,177"/><area shape="rect" id="node15" href="classhivpopulation.html#a299863ae882da177d548edc26d9a2cb8" title="Construct a HIV population with certain parameters." alt="" coords="597,201,776,228"/></map>
</div>
</p>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a class="anchor" id="a2929fa585f451eff209d5766e280d0b6"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double* haploid_highd::allele_frequencies</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="aa496b4cc82e39329cc6d18b1432cd515"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::carrying_capacity</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a1f25e3ce87377569a3b14c92bb72fc55"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double* haploid_highd::chi1</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a0047d21da6fe6fa9c1ce94b7b53ae37a"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double** haploid_highd::chi2</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="abc8e66940a4dd00a25388ec322cf7a49"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool haploid_highd::circular</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a023aab1a8360b632e471bdc3a4ec37d1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::crossover_rate</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a495c975322ea03d01417be29e951a1f0"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int* haploid_highd::crossovers</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a7b2a100e6109b0190f3435b08097c327"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt;<a class="el" href="structclone__t.html">clone_t</a>&gt;* haploid_highd::current_pop</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="afe01d72044a8d81a32694287aa68ef09"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">gsl_rng* haploid_highd::evo_generator</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a1a57b48e7395ef197c38c46329af820a"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::fitness_max</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="aa3cd0c02607b41b378f7fa488d5b5b46"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structstat__t.html">stat_t</a> haploid_highd::fitness_stat</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="abcade2de00cd6efee795651d46eede3a"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double* haploid_highd::gamete_allele_frequencies</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a771e8b659f399927ed612ba2f01c7ef8"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::generation</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a8a44ffb375d6763ebdef28d9e5121d8d"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int* haploid_highd::genome</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a9395ef40e6478fbc361c82193377346e"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">gsl_rng* haploid_highd::label_generator</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ad55b1105f6869684863c7e4f34023827"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::mutation_rate</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ad0f15c886bb03789063f5587e080166e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt;<a class="el" href="structclone__t.html">clone_t</a>&gt;* haploid_highd::new_pop</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ad1a76259095c3119e33fcc63abd52227"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::number_of_loci</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="af9bef51d5530a1807d79063fcaf81363"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::number_of_traits</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a4dc7a29976c59a5c4755d12ade075069"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::outcrossing_rate</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ac0c344e3d804d8f788ea527167a9bddc"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::outcrossing_rate_effective</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ae57c154a248e0cfe9be446a04d1765d5"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double haploid_highd::participation_ratio</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a13f8881a0a81b4cc1ac3fe23e95c27fd"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::population_size</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a1c647247fb891a314caf21b58f5a1e0d"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt;int&gt; haploid_highd::random_sample</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="afa9a75efb02352d951e633f31ac1bc59"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::recombination_model</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ad364734ea945abb84658303eefa0d817"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::scratch</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a1d7ee5c309f75b8b5d97f087569c532d"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int haploid_highd::seed</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a720e92ddf95f0aaf1e5136b60b53caad"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt;int&gt; haploid_highd::sex_gametes</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a3faa49df0828300313fb979e29b53bd8"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classhypercube__highd.html">hypercube_highd</a>* haploid_highd::trait</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a870cd9eb97be85b79de12baba7231aa7"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double** haploid_highd::trait_covariance</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a97fe03c6b3b3034410d728f17fa2716b"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structstat__t.html">stat_t</a>* haploid_highd::trait_stat</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="ffpopsim__highd_8h_source.html">ffpopsim_highd.h</a></li>
<li>src/<a class="el" href="haploid__highd_8cpp.html">haploid_highd.cpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="classhaploid__highd.html">haploid_highd</a></li>
    <li class="footer">Generated on Fri Aug 31 2012 16:53:16 for FFPopSim by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.2 </li>
  </ul>
</div>
</body>
</html>
